--- linux-source-2.6.12/include/asm-arm/arch-ixp4xx/timex.h~	2005-06-18 05:18:29.000000000 +0930
+++ linux-source-2.6.12/include/asm-arm/arch-ixp4xx/timex.h	2005-10-07 10:34:54.957442736 +0930
@@ -8,6 +8,10 @@
 /*
  * We use IXP425 General purpose timer for our timer needs, it runs at 
  * 66.66... MHz
+ * The NSLU2 has a 33.00MHz crystal, so a different CLOCK_TICK_RATE is required.
  */
+#ifdef CONFIG_MACH_NSLU2
+#define CLOCK_TICK_RATE (66000000)
+#else
 #define CLOCK_TICK_RATE (66666666)
-
+#endif
