Index: apex-1.4.11/src/mach-ixp42x/debian-nas100d-arm_config
===================================================================
--- apex-1.4.11.orig/src/mach-ixp42x/debian-nas100d-arm_config
+++ apex-1.4.11/src/mach-ixp42x/debian-nas100d-arm_config
@@ -1,7 +1,7 @@
 #
 # Automatically generated make config: don't edit
 # Linux kernel version: 
-# Mon Jan  8 23:59:24 2007
+# Tue Jan  9 22:11:52 2007
 #
 CONFIG_ARM=y
 CONFIG_CPU_ARMV5=y
@@ -110,7 +110,8 @@ CONFIG_ENV_MUTABLE=y
 CONFIG_ENV_LINK=y
 CONFIG_ENV_REGION="nor:0xee0000+15k"
 CONFIG_ENV_CHECK_LEN=1024
-# CONFIG_VARIATIONS is not set
+CONFIG_VARIATIONS=y
+CONFIG_VARIATION_SUFFIX="-alt"
 # CONFIG_NO_BOOTSTRAP is not set
 CONFIG_NOR_BOOT=y
 # CONFIG_NAND_BOOT is not set
@@ -131,14 +132,18 @@ CONFIG_ENV_STARTUP_RAMDISK_COPY=y
 #    Regions
 #
 CONFIG_ENV_REGION_KERNEL="fis://kernel"
+# CONFIG_ENV_REGION_KERNEL_ALT_P is not set
 CONFIG_ENV_REGION_RAMDISK="fis://filesystem"
+# CONFIG_ENV_REGION_RAMDISK_ALT_P is not set
 
 #
 #    Overrides
 #
 CONFIG_ENV_DEFAULT_CMDLINE_OVERRIDE=y
 CONFIG_ENV_DEFAULT_CMDLINE="console=ttyS0,115200 rtc-pcf8563.probe=0,0x51 noirqdebug"
+# CONFIG_ENV_DEFAULT_CMDLINE_ALT_P is not set
 # CONFIG_ENV_DEFAULT_STARTUP_OVERRIDE is not set
+# CONFIG_ENV_DEFAULT_STARTUP_ALT_P is not set
 CONFIG_USES_NOR_BOOTFLASH=y
 CONFIG_LITTLEENDIAN=y
 CONFIG_CLEAR_STACKS=y
Index: apex-1.4.11/src/mach-ixp42x/debian-nas100d-armeb_config
===================================================================
--- apex-1.4.11.orig/src/mach-ixp42x/debian-nas100d-armeb_config
+++ apex-1.4.11/src/mach-ixp42x/debian-nas100d-armeb_config
@@ -110,7 +110,7 @@ CONFIG_ENV_MUTABLE=y
 CONFIG_ENV_LINK=y
 CONFIG_ENV_REGION="nor:0xee0000+15k"
 CONFIG_ENV_CHECK_LEN=1024
-# CONFIG_VARIATIONS is not set
+CONFIG_VARIATIONS=y
 # CONFIG_NO_BOOTSTRAP is not set
 CONFIG_NOR_BOOT=y
 # CONFIG_NAND_BOOT is not set
Index: apex-1.4.11/src/mach-ixp42x/debian-nslu2-arm_config
===================================================================
--- apex-1.4.11.orig/src/mach-ixp42x/debian-nslu2-arm_config
+++ apex-1.4.11/src/mach-ixp42x/debian-nslu2-arm_config
@@ -1,7 +1,7 @@
 #
 # Automatically generated make config: don't edit
 # Linux kernel version: 
-# Mon Jan  8 23:59:24 2007
+# Tue Jan  9 22:11:09 2007
 #
 CONFIG_ARM=y
 CONFIG_CPU_ARMV5=y
@@ -112,7 +112,8 @@ CONFIG_ENV_MUTABLE=y
 CONFIG_ENV_LINK=y
 CONFIG_ENV_REGION="nor:0x7c000+15k"
 CONFIG_ENV_CHECK_LEN=1024
-# CONFIG_VARIATIONS is not set
+CONFIG_VARIATIONS=y
+CONFIG_VARIATION_SUFFIX="-alt"
 # CONFIG_NO_BOOTSTRAP is not set
 CONFIG_NOR_BOOT=y
 # CONFIG_NAND_BOOT is not set
@@ -133,14 +134,18 @@ CONFIG_ENV_REGION_RAMDISK_SWAP=y
 #    Regions
 #
 CONFIG_ENV_REGION_KERNEL="fis://kernel"
+# CONFIG_ENV_REGION_KERNEL_ALT_P is not set
 CONFIG_ENV_REGION_RAMDISK="fis://ramdisk"
+# CONFIG_ENV_REGION_RAMDISK_ALT_P is not set
 
 #
 #    Overrides
 #
 CONFIG_ENV_DEFAULT_CMDLINE_OVERRIDE=y
 CONFIG_ENV_DEFAULT_CMDLINE="console=ttyS0,115200 rtc-x1205.probe=0,0x6f noirqdebug"
+# CONFIG_ENV_DEFAULT_CMDLINE_ALT_P is not set
 # CONFIG_ENV_DEFAULT_STARTUP_OVERRIDE is not set
+# CONFIG_ENV_DEFAULT_STARTUP_ALT_P is not set
 CONFIG_USES_NOR_BOOTFLASH=y
 CONFIG_LITTLEENDIAN=y
 CONFIG_CLEAR_STACKS=y
Index: apex-1.4.11/src/mach-ixp42x/debian-nslu2-armeb_config
===================================================================
--- apex-1.4.11.orig/src/mach-ixp42x/debian-nslu2-armeb_config
+++ apex-1.4.11/src/mach-ixp42x/debian-nslu2-armeb_config
@@ -112,7 +112,7 @@ CONFIG_ENV_MUTABLE=y
 CONFIG_ENV_LINK=y
 CONFIG_ENV_REGION="nor:0x7fc000+15k"
 CONFIG_ENV_CHECK_LEN=1024
-# CONFIG_VARIATIONS is not set
+CONFIG_VARIATIONS=y
 # CONFIG_NO_BOOTSTRAP is not set
 CONFIG_NOR_BOOT=y
 # CONFIG_NAND_BOOT is not set
Index: apex-1.4.11/src/mach-ixp42x/nslu2_config
===================================================================
--- apex-1.4.11.orig/src/mach-ixp42x/nslu2_config
+++ apex-1.4.11/src/mach-ixp42x/nslu2_config
@@ -108,7 +108,7 @@ CONFIG_ENV_MUTABLE=y
 CONFIG_ENV_LINK=y
 CONFIG_ENV_REGION="nor:128k+64k"
 CONFIG_ENV_CHECK_LEN=1024
-# CONFIG_VARIATIONS is not set
+CONFIG_VARIATIONS=y
 # CONFIG_NO_BOOTSTRAP is not set
 CONFIG_NOR_BOOT=y
 # CONFIG_NAND_BOOT is not set
Index: apex-1.4.11/src/mach-ixp42x/slugos-dsmg600-arm_config
===================================================================
--- apex-1.4.11.orig/src/mach-ixp42x/slugos-dsmg600-arm_config
+++ apex-1.4.11/src/mach-ixp42x/slugos-dsmg600-arm_config
@@ -108,7 +108,7 @@ CONFIG_ENV_MUTABLE=y
 CONFIG_ENV_LINK=y
 CONFIG_ENV_REGION="nor:0xee0000+15k"
 CONFIG_ENV_CHECK_LEN=1024
-# CONFIG_VARIATIONS is not set
+CONFIG_VARIATIONS=y
 # CONFIG_NO_BOOTSTRAP is not set
 CONFIG_NOR_BOOT=y
 # CONFIG_NAND_BOOT is not set
Index: apex-1.4.11/src/mach-ixp42x/slugos-dsmg600-armeb_config
===================================================================
--- apex-1.4.11.orig/src/mach-ixp42x/slugos-dsmg600-armeb_config
+++ apex-1.4.11/src/mach-ixp42x/slugos-dsmg600-armeb_config
@@ -108,7 +108,7 @@ CONFIG_ENV_MUTABLE=y
 CONFIG_ENV_LINK=y
 CONFIG_ENV_REGION="nor:0xee0000+15k"
 CONFIG_ENV_CHECK_LEN=1024
-# CONFIG_VARIATIONS is not set
+CONFIG_VARIATIONS=y
 # CONFIG_NO_BOOTSTRAP is not set
 CONFIG_NOR_BOOT=y
 # CONFIG_NAND_BOOT is not set
Index: apex-1.4.11/src/mach-ixp42x/slugos-nas100d-arm_config
===================================================================
--- apex-1.4.11.orig/src/mach-ixp42x/slugos-nas100d-arm_config
+++ apex-1.4.11/src/mach-ixp42x/slugos-nas100d-arm_config
@@ -108,7 +108,7 @@ CONFIG_ENV_MUTABLE=y
 CONFIG_ENV_LINK=y
 CONFIG_ENV_REGION="nor:0xee0000+15k"
 CONFIG_ENV_CHECK_LEN=1024
-# CONFIG_VARIATIONS is not set
+CONFIG_VARIATIONS=y
 # CONFIG_NO_BOOTSTRAP is not set
 CONFIG_NOR_BOOT=y
 # CONFIG_NAND_BOOT is not set
Index: apex-1.4.11/src/mach-ixp42x/slugos-nas100d-armeb_config
===================================================================
--- apex-1.4.11.orig/src/mach-ixp42x/slugos-nas100d-armeb_config
+++ apex-1.4.11/src/mach-ixp42x/slugos-nas100d-armeb_config
@@ -1,7 +1,7 @@
 #
 # Automatically generated make config: don't edit
 # Linux kernel version: 
-# Mon Jan  8 23:59:25 2007
+# Tue Jan  9 22:08:07 2007
 #
 CONFIG_ARM=y
 CONFIG_CPU_ARMV5=y
@@ -108,7 +108,8 @@ CONFIG_ENV_MUTABLE=y
 CONFIG_ENV_LINK=y
 CONFIG_ENV_REGION="nor:0xee0000+15k"
 CONFIG_ENV_CHECK_LEN=1024
-# CONFIG_VARIATIONS is not set
+CONFIG_VARIATIONS=y
+CONFIG_VARIATION_SUFFIX="-alt"
 # CONFIG_NO_BOOTSTRAP is not set
 CONFIG_NOR_BOOT=y
 # CONFIG_NAND_BOOT is not set
@@ -127,13 +128,16 @@ CONFIG_ENV_STARTUP_KERNEL_COPY=y
 #    Regions
 #
 CONFIG_ENV_REGION_KERNEL="fis://kernel"
+# CONFIG_ENV_REGION_KERNEL_ALT_P is not set
 
 #
 #    Overrides
 #
 CONFIG_ENV_DEFAULT_CMDLINE_OVERRIDE=y
 CONFIG_ENV_DEFAULT_CMDLINE="root=/dev/mtdblock2 rootfstype=jffs2 console=ttyS0,115200"
+# CONFIG_ENV_DEFAULT_CMDLINE_ALT_P is not set
 # CONFIG_ENV_DEFAULT_STARTUP_OVERRIDE is not set
+# CONFIG_ENV_DEFAULT_STARTUP_ALT_P is not set
 CONFIG_USES_NOR_BOOTFLASH=y
 CONFIG_BIGENDIAN=y
 CONFIG_CLEAR_STACKS=y
Index: apex-1.4.11/src/mach-ixp42x/slugos-nslu2-arm_config
===================================================================
--- apex-1.4.11.orig/src/mach-ixp42x/slugos-nslu2-arm_config
+++ apex-1.4.11/src/mach-ixp42x/slugos-nslu2-arm_config
@@ -110,7 +110,7 @@ CONFIG_ENV_MUTABLE=y
 CONFIG_ENV_LINK=y
 CONFIG_ENV_REGION="nor:0x7f8000+16k"
 CONFIG_ENV_CHECK_LEN=1024
-# CONFIG_VARIATIONS is not set
+CONFIG_VARIATIONS=y
 # CONFIG_NO_BOOTSTRAP is not set
 CONFIG_NOR_BOOT=y
 # CONFIG_NAND_BOOT is not set
Index: apex-1.4.11/src/mach-ixp42x/slugos-nslu2-armeb_config
===================================================================
--- apex-1.4.11.orig/src/mach-ixp42x/slugos-nslu2-armeb_config
+++ apex-1.4.11/src/mach-ixp42x/slugos-nslu2-armeb_config
@@ -1,7 +1,7 @@
 #
 # Automatically generated make config: don't edit
 # Linux kernel version: 
-# Mon Jan  8 23:59:27 2007
+# Tue Jan  9 22:10:03 2007
 #
 CONFIG_ARM=y
 CONFIG_CPU_ARMV5=y
@@ -110,7 +110,8 @@ CONFIG_ENV_MUTABLE=y
 CONFIG_ENV_LINK=y
 CONFIG_ENV_REGION="nor:0x7f8000+16k"
 CONFIG_ENV_CHECK_LEN=1024
-# CONFIG_VARIATIONS is not set
+CONFIG_VARIATIONS=y
+CONFIG_VARIATION_SUFFIX="-alt"
 # CONFIG_NO_BOOTSTRAP is not set
 CONFIG_NOR_BOOT=y
 # CONFIG_NAND_BOOT is not set
@@ -129,13 +130,16 @@ CONFIG_ENV_STARTUP_KERNEL_COPY=y
 #    Regions
 #
 CONFIG_ENV_REGION_KERNEL="fis://kernel"
+# CONFIG_ENV_REGION_KERNEL_ALT_P is not set
 
 #
 #    Overrides
 #
 CONFIG_ENV_DEFAULT_CMDLINE_OVERRIDE=y
 CONFIG_ENV_DEFAULT_CMDLINE="root=/dev/mtdblock4 rootfstype=jffs2 console=ttyS0,115200"
+# CONFIG_ENV_DEFAULT_CMDLINE_ALT_P is not set
 # CONFIG_ENV_DEFAULT_STARTUP_OVERRIDE is not set
+# CONFIG_ENV_DEFAULT_STARTUP_ALT_P is not set
 CONFIG_USES_NOR_BOOTFLASH=y
 CONFIG_BIGENDIAN=y
 CONFIG_CLEAR_STACKS=y
