From: Rod Whitby <rod@whitby.id.au>
Subject: Enable memtest command on ixp4xx machines

Signed-off-by: Rod Whitby <rod@whitby.id.au>
--

Index: apex/src/mach-ixp42x/debian-nslu2-arm_config
===================================================================
--- apex.orig/src/mach-ixp42x/debian-nslu2-arm_config	2008-05-28 10:52:50.000000000 +0930
+++ apex/src/mach-ixp42x/debian-nslu2-arm_config	2008-05-28 10:53:00.000000000 +0930
@@ -92,7 +92,7 @@
 CONFIG_CMD_XRECEIVE=y
 CONFIG_CMD_MEMLIMIT=y
 CONFIG_CMD_MEMSCAN=y
-# CONFIG_CMD_MEMTEST is not set
+CONFIG_CMD_MEMTEST=y
 
 #
 # Generic Drivers
Index: apex/src/mach-ixp42x/debian-nslu2-armeb_config
===================================================================
--- apex.orig/src/mach-ixp42x/debian-nslu2-armeb_config	2008-05-28 10:52:50.000000000 +0930
+++ apex/src/mach-ixp42x/debian-nslu2-armeb_config	2008-05-28 10:53:00.000000000 +0930
@@ -92,7 +92,7 @@
 CONFIG_CMD_XRECEIVE=y
 CONFIG_CMD_MEMLIMIT=y
 CONFIG_CMD_MEMSCAN=y
-# CONFIG_CMD_MEMTEST is not set
+CONFIG_CMD_MEMTEST=y
 
 #
 # Generic Drivers
Index: apex/src/mach-ixp42x/nslu2_config
===================================================================
--- apex.orig/src/mach-ixp42x/nslu2_config	2008-05-28 10:52:50.000000000 +0930
+++ apex/src/mach-ixp42x/nslu2_config	2008-05-28 10:53:00.000000000 +0930
@@ -90,7 +90,7 @@
 CONFIG_CMD_XRECEIVE=y
 CONFIG_CMD_MEMLIMIT=y
 CONFIG_CMD_MEMSCAN=y
-# CONFIG_CMD_MEMTEST is not set
+CONFIG_CMD_MEMTEST=y
 
 #
 # Generic Drivers
Index: apex/src/mach-ixp42x/slugos-nslu2-arm_config
===================================================================
--- apex.orig/src/mach-ixp42x/slugos-nslu2-arm_config	2008-05-28 10:52:58.000000000 +0930
+++ apex/src/mach-ixp42x/slugos-nslu2-arm_config	2008-05-28 10:53:00.000000000 +0930
@@ -90,7 +90,7 @@
 CONFIG_CMD_XRECEIVE=y
 CONFIG_CMD_MEMLIMIT=y
 CONFIG_CMD_MEMSCAN=y
-# CONFIG_CMD_MEMTEST is not set
+CONFIG_CMD_MEMTEST=y
 
 #
 # Generic Drivers
Index: apex/src/mach-ixp42x/slugos-nslu2-armeb_config
===================================================================
--- apex.orig/src/mach-ixp42x/slugos-nslu2-armeb_config	2008-05-28 10:52:50.000000000 +0930
+++ apex/src/mach-ixp42x/slugos-nslu2-armeb_config	2008-05-28 10:53:00.000000000 +0930
@@ -90,7 +90,7 @@
 CONFIG_CMD_XRECEIVE=y
 CONFIG_CMD_MEMLIMIT=y
 CONFIG_CMD_MEMSCAN=y
-# CONFIG_CMD_MEMTEST is not set
+CONFIG_CMD_MEMTEST=y
 
 #
 # Generic Drivers
