From: Rod Whitby <rod@whitby.id.au>
Subject: Fix apex location for big-endian Debian on NSLU2

Move the apex environment into the apex partition.

Signed-off-by: Rod Whitby <rod@whitby.id.au>
--

Index: apex/src/mach-ixp42x/debian-nslu2-armeb_config
===================================================================
--- apex.orig/src/mach-ixp42x/debian-nslu2-armeb_config	2008-05-28 10:53:00.000000000 +0930
+++ apex/src/mach-ixp42x/debian-nslu2-armeb_config	2008-05-28 10:53:02.000000000 +0930
@@ -119,7 +119,7 @@
 CONFIG_ENV=y
 CONFIG_ENV_LINK=y
 CONFIG_ENV_MUTABLE=y
-CONFIG_ENV_REGION="nor:0x7fc000+15k"
+CONFIG_ENV_REGION="nor:0x7c000+15k"
 # CONFIG_ENV_SAVEATONCE is not set
 CONFIG_ENV_CHECK_LEN=1024
 CONFIG_VARIATIONS=y
