Index: apex-1.5.13/src/mach-ixp42x/Makefile
===================================================================
--- apex-1.5.13.orig/src/mach-ixp42x/Makefile	2007-02-25 08:23:43.000000000 +1030
+++ apex-1.5.13/src/mach-ixp42x/Makefile	2008-01-02 20:43:50.000000000 +1030
@@ -48,6 +48,7 @@
 
 obj-x$(CONFIG_MACH_NSLU2)$(CONFIG_SPINNER) += spinner-nslu2.o
 obj-x$(CONFIG_MACH_NSLU2)$(CONFIG_VARIATIONS) += variation-nslu2.o
+obj-x$(CONFIG_MACH_FSG3)$(CONFIG_VARIATIONS) += variation-fsg3.o
 obj-$(CONFIG_INTERRUPTS)	+= exception_vectors.o
 
 obj-$(CONFIG_NPE)		+= npe.o npe-interface.o
Index: apex-1.5.13/src/mach-ixp42x/variation-fsg3.c
===================================================================
--- /dev/null	1970-01-01 00:00:00.000000000 +0000
+++ apex-1.5.13/src/mach-ixp42x/variation-fsg3.c	2008-01-02 20:43:50.000000000 +1030
@@ -0,0 +1,40 @@
+/* variation-fsg3.c
+
+   written by Rod Whitby
+   2 Jan 2008
+
+   based on variation-nslu2.c
+   written by Marc Singer
+   6 Jan 2007
+
+   Copyright (C) 2007 Marc Singer
+   Copyright (C) 2008 Rod Whitby
+
+   -----------
+   DESCRIPTION
+   -----------
+
+   Service to support variable variations on the fsg3.  Essetially,
+   this is the interface with the user to select the variation at
+   boot-time.
+
+*/
+
+#include <config.h>
+#include <alias.h>
+#include <service.h>
+#include <apex.h>
+
+#include "hardware.h"
+
+static void variation_init (void)
+{
+  int button_pressed = (GPIO_INR & (1<<GPIO_I_SYNCBUTTON)) == 0;
+
+  if (button_pressed)
+    alias_set ("variation", CONFIG_VARIATION_SUFFIX);
+}
+
+static __service_9 struct service_d variation_service = {
+  .init = variation_init,
+};
Index: apex-1.5.13/src/mach-ixp42x/slugos-fsg3-armeb_config
===================================================================
--- apex-1.5.13.orig/src/mach-ixp42x/slugos-fsg3-armeb_config	2008-01-02 20:43:50.000000000 +1030
+++ apex-1.5.13/src/mach-ixp42x/slugos-fsg3-armeb_config	2008-01-02 20:43:50.000000000 +1030
@@ -118,7 +118,8 @@
 CONFIG_ENV_REGION="nor:0x3bc000+15k"
 # CONFIG_ENV_SAVEATONCE is not set
 CONFIG_ENV_CHECK_LEN=1024
-# CONFIG_VARIATIONS is not set
+CONFIG_VARIATIONS=y
+CONFIG_VARIATION_SUFFIX="-alt"
 # CONFIG_NO_BOOTSTRAP is not set
 CONFIG_NOR_BOOT=y
 # CONFIG_NAND_BOOT is not set
@@ -139,13 +140,18 @@
 #    Regions
 #
 CONFIG_ENV_REGION_KERNEL="fis://kern1"
+CONFIG_ENV_REGION_KERNEL_ALT_P=y
+CONFIG_ENV_REGION_KERNEL_ALT="fis://kern2"
 
 #
 #    Overrides
 #
 CONFIG_ENV_DEFAULT_CMDLINE_OVERRIDE=y
-CONFIG_ENV_DEFAULT_CMDLINE="root=/dev/hda1 console=ttyS0,115200"
+CONFIG_ENV_DEFAULT_CMDLINE="root=/dev/hda1 rootdelay=10 console=ttyS0,115200"
+CONFIG_ENV_DEFAULT_CMDLINE_ALT_P=y
+CONFIG_ENV_DEFAULT_CMDLINE_ALT="root=/dev/hda2 rootdelay=10 console=ttyS0,115200"
 # CONFIG_ENV_DEFAULT_STARTUP_OVERRIDE is not set
+# CONFIG_ENV_DEFAULT_STARTUP_ALT_P is not set
 CONFIG_USES_NOR_BOOTFLASH=y
 CONFIG_RELOCATE_SIMPLE=y
 CONFIG_INLINE_PLATFORM_INITIALIZATION=y
Index: apex-1.5.13/src/mach-ixp42x/slugos-fsg3-arm_config
===================================================================
--- apex-1.5.13.orig/src/mach-ixp42x/slugos-fsg3-arm_config	2008-01-02 22:05:07.000000000 +1030
+++ apex-1.5.13/src/mach-ixp42x/slugos-fsg3-arm_config	2008-01-02 22:05:09.000000000 +1030
@@ -118,7 +118,8 @@
 CONFIG_ENV_REGION="nor:0x3bc000+15k"
 # CONFIG_ENV_SAVEATONCE is not set
 CONFIG_ENV_CHECK_LEN=1024
-# CONFIG_VARIATIONS is not set
+CONFIG_VARIATIONS=y
+CONFIG_VARIATION_SUFFIX="-alt"
 # CONFIG_NO_BOOTSTRAP is not set
 CONFIG_NOR_BOOT=y
 # CONFIG_NAND_BOOT is not set
@@ -139,13 +140,18 @@
 #    Regions
 #
 CONFIG_ENV_REGION_KERNEL="fis://kern1"
+CONFIG_ENV_REGION_KERNEL_ALT_P=y
+CONFIG_ENV_REGION_KERNEL_ALT="fis://kern2"
 
 #
 #    Overrides
 #
 CONFIG_ENV_DEFAULT_CMDLINE_OVERRIDE=y
-CONFIG_ENV_DEFAULT_CMDLINE="root=/dev/hda1 console=ttyS0,115200"
+CONFIG_ENV_DEFAULT_CMDLINE="root=/dev/hda1 rootdelay=10 console=ttyS0,115200"
+CONFIG_ENV_DEFAULT_CMDLINE_ALT_P=y
+CONFIG_ENV_DEFAULT_CMDLINE_ALT="root=/dev/hda2 rootdelay=10 console=ttyS0,115200"
 # CONFIG_ENV_DEFAULT_STARTUP_OVERRIDE is not set
+# CONFIG_ENV_DEFAULT_STARTUP_ALT_P is not set
 CONFIG_USES_NOR_BOOTFLASH=y
 CONFIG_RELOCATE_SIMPLE=y
 CONFIG_INLINE_PLATFORM_INITIALIZATION=y
