Index: linux-2.6.20-rc1/arch/arm/mach-ixp4xx/avila-setup.c
===================================================================
--- linux-2.6.20-rc1.orig/arch/arm/mach-ixp4xx/avila-setup.c	2006-12-17 16:53:51.972719710 +0000
+++ linux-2.6.20-rc1/arch/arm/mach-ixp4xx/avila-setup.c	2006-12-17 16:54:27.579603463 +0000
@@ -104,6 +104,16 @@
 	.resource		= avila_uart_resources
 };
 
+struct npe_ucode_platform_data avila_npe_ucode_data = {
+	 .mtd_partition  = "RedBoot",
+};
+
+static struct platform_device avila_npe_ucode = {
+       .name                   = "ixp4xx_npe_ucode",
+       .id                     = 0,
+       .dev.platform_data      = &avila_npe_ucode_data,
+};
+
 static struct resource avila_pata_resources[] = {
 	{
 		.flags	= IORESOURCE_MEM
@@ -183,7 +193,7 @@
 	&avila_uart,
 	&mac0,
 	&mac1,
-
+	&avila_npe_ucode,
 };
 
 static void __init avila_init(void)
