Index: linux-2.6.22-rc1-armeb/include/asm-arm/arch-ixp4xx/ixp4xx-regs.h
===================================================================
--- linux-2.6.22-rc1-armeb.orig/include/asm-arm/arch-ixp4xx/ixp4xx-regs.h	2007-05-16 07:33:34.000000000 -0700
+++ linux-2.6.22-rc1-armeb/include/asm-arm/arch-ixp4xx/ixp4xx-regs.h	2007-05-16 07:51:08.000000000 -0700
@@ -607,4 +607,19 @@
 
 #define DCMD_LENGTH	0x01fff		/* length mask (max = 8K - 1) */
 
+#ifndef __ASSEMBLY__
+static inline int cpu_is_ixp46x(void)
+{
+#ifdef CONFIG_CPU_IXP46X
+	unsigned int processor_id;
+
+	asm("mrc p15, 0, %0, cr0, cr0, 0;" : "=r"(processor_id) :);
+
+	if ((processor_id & 0xffffff00) == 0x69054200)
+		return 1;
+#endif
+	return 0;
+}
+#endif
+
 #endif
Index: linux-2.6.22-rc1-armeb/include/asm-arm/arch-ixp4xx/cpu.h
===================================================================
--- linux-2.6.22-rc1-armeb.orig/include/asm-arm/arch-ixp4xx/cpu.h	2007-05-16 05:55:08.000000000 -0700
+++ /dev/null	1970-01-01 00:00:00.000000000 +0000
@@ -1,31 +0,0 @@
-/*
- * include/asm-arm/arch-ixp4xx/cpu.h
- *
- * IXP4XX cpu type detection
- *
- * Copyright (C) 2007 MontaVista Software, Inc.
- *
- * This program is free software; you can redistribute it and/or modify
- * it under the terms of the GNU General Public License version 2 as
- * published by the Free Software Foundation.
- *
- */
-
-#ifndef __ASM_ARCH_CPU_H__
-#define __ASM_ARCH_CPU_H__
-
-extern unsigned int processor_id;
-/* Processor id value in CP15 Register 0 */
-#define IXP425_PROCESSOR_ID_VALUE	0x690541c0
-#define IXP435_PROCESSOR_ID_VALUE	0x69054040
-#define IXP465_PROCESSOR_ID_VALUE	0x69054200
-#define IXP4XX_PROCESSOR_ID_MASK	0xfffffff0
-
-#define cpu_is_ixp42x()	((processor_id & IXP4XX_PROCESSOR_ID_MASK) == \
-			  IXP425_PROCESSOR_ID_VALUE)
-#define cpu_is_ixp43x()	((processor_id & IXP4XX_PROCESSOR_ID_MASK) == \
-			  IXP435_PROCESSOR_ID_VALUE)
-#define cpu_is_ixp46x()	((processor_id & IXP4XX_PROCESSOR_ID_MASK) == \
-			  IXP465_PROCESSOR_ID_VALUE)
-
-#endif  /* _ASM_ARCH_CPU_H */
Index: linux-2.6.22-rc1-armeb/include/asm-arm/arch-ixp4xx/hardware.h
===================================================================
--- linux-2.6.22-rc1-armeb.orig/include/asm-arm/arch-ixp4xx/hardware.h	2007-05-16 07:33:38.000000000 -0700
+++ linux-2.6.22-rc1-armeb/include/asm-arm/arch-ixp4xx/hardware.h	2007-05-16 07:51:08.000000000 -0700
@@ -17,8 +17,8 @@
 #ifndef __ASM_ARCH_HARDWARE_H__
 #define __ASM_ARCH_HARDWARE_H__
 
-#define PCIBIOS_MIN_IO		0x00001000
-#define PCIBIOS_MIN_MEM		(cpu_is_ixp43x() ? 0x40000000 : 0x48000000)
+#define PCIBIOS_MIN_IO			0x00001000
+#define PCIBIOS_MIN_MEM			0x48000000
 
 /*
  * We override the standard dma-mask routines for bouncing.
@@ -27,8 +27,11 @@
 
 #define pcibios_assign_all_busses()	1
 
-#ifndef __ASSEMBLER__
-#include <asm/arch/cpu.h>
+#if defined(CONFIG_CPU_IXP46X) && !defined(__ASSEMBLY__)
+extern unsigned int processor_id;
+#define cpu_is_ixp465() ((processor_id & 0xffffffc0) == 0x69054200)
+#else
+#define	cpu_is_ixp465()	(0)
 #endif
 
 /* Register locations and bits */
Index: linux-2.6.22-rc1-armeb/include/asm-arm/arch-ixp4xx/irqs.h
===================================================================
--- linux-2.6.22-rc1-armeb.orig/include/asm-arm/arch-ixp4xx/irqs.h	2007-05-16 07:48:53.000000000 -0700
+++ linux-2.6.22-rc1-armeb/include/asm-arm/arch-ixp4xx/irqs.h	2007-05-16 07:51:08.000000000 -0700
@@ -62,10 +62,10 @@
 /*
  * Only first 32 sources are valid if running on IXP42x systems
  */
-#if defined(CONFIG_CPU_IXP46X) || defined(CONFIG_CPU_IXP43X)
-#define NR_IRQS			64
-#else
+#ifndef	CONFIG_CPU_IXP46X
 #define NR_IRQS			32
+#else
+#define NR_IRQS			64
 #endif
 
 #define	XSCALE_PMU_IRQ		(IRQ_IXP4XX_XSCALE_PMU)
Index: linux-2.6.22-rc1-armeb/arch/arm/mach-ixp4xx/common-pci.c
===================================================================
--- linux-2.6.22-rc1-armeb.orig/arch/arm/mach-ixp4xx/common-pci.c	2007-05-16 05:54:59.000000000 -0700
+++ linux-2.6.22-rc1-armeb/arch/arm/mach-ixp4xx/common-pci.c	2007-05-16 07:51:08.000000000 -0700
@@ -374,7 +374,7 @@
 	 * Determine which PCI read method to use.
 	 * Rev 0 IXP425 requires workaround.
 	 */
-	if (!(processor_id & 0xf) && cpu_is_ixp42x()) {
+	if (!(processor_id & 0xf) && !cpu_is_ixp46x()) {
 		printk("PCI: IXP42x A0 silicon detected - "
 			"PCI Non-Prefetch Workaround Enabled\n");
 		ixp4xx_pci_read = ixp4xx_pci_read_errata;
@@ -480,7 +480,7 @@
 	res[0].flags = IORESOURCE_IO;
 
 	res[1].name = "PCI Memory Space";
-	res[1].start = PCIBIOS_MIN_MEM;
+	res[1].start = 0x48000000;
 #ifndef CONFIG_IXP4XX_INDIRECT_PCI
 	res[1].end = 0x4bffffff;
 #else
Index: linux-2.6.22-rc1-armeb/arch/arm/mm/proc-xscale.S
===================================================================
--- linux-2.6.22-rc1-armeb.orig/arch/arm/mm/proc-xscale.S	2007-05-16 05:54:59.000000000 -0700
+++ linux-2.6.22-rc1-armeb/arch/arm/mm/proc-xscale.S	2007-05-16 07:51:08.000000000 -0700
@@ -584,11 +584,6 @@
 	.asciz	"XScale-IXP42x Family"
 	.size	cpu_ixp42x_name, . - cpu_ixp42x_name
 
-	.type	cpu_ixp43x_name, #object
-cpu_ixp43x_name:
-	.asciz	"XScale-IXP43x Family"
-	.size	cpu_ixp43x_name, . - cpu_ixp43x_name
-
 	.type	cpu_ixp46x_name, #object
 cpu_ixp46x_name:
 	.asciz	"XScale-IXP46x Family"
@@ -848,29 +843,6 @@
 	.long	xscale_cache_fns
 	.size   __ixp42x_proc_info, . - __ixp42x_proc_info                
 
-	.type   __ixp43x_proc_info, #object
-__ixp43x_proc_info:
-	.long   0x69054040
-	.long   0xfffffff0
-	.long   PMD_TYPE_SECT | \
-		PMD_SECT_BUFFERABLE | \
-		PMD_SECT_CACHEABLE | \
-		PMD_SECT_AP_WRITE | \
-		PMD_SECT_AP_READ
-	.long   PMD_TYPE_SECT | \
-		PMD_SECT_AP_WRITE | \
-		PMD_SECT_AP_READ
-	b       __xscale_setup
-	.long   cpu_arch_name
-	.long   cpu_elf_name
-	.long   HWCAP_SWP|HWCAP_HALF|HWCAP_THUMB|HWCAP_FAST_MULT|HWCAP_EDSP
-	.long   cpu_ixp43x_name
-	.long   xscale_processor_functions
-	.long   v4wbi_tlb_fns
-	.long   xscale_mc_user_fns
-	.long   xscale_cache_fns
-	.size   __ixp43x_proc_info, . - __ixp43x_proc_info
-
 	.type	__ixp46x_proc_info, #object
 __ixp46x_proc_info:
 	.long   0x69054200
Index: linux-2.6.22-rc1-armeb/arch/arm/mach-ixp4xx/common.c
===================================================================
--- linux-2.6.22-rc1-armeb.orig/arch/arm/mach-ixp4xx/common.c	2007-05-16 07:48:53.000000000 -0700
+++ linux-2.6.22-rc1-armeb/arch/arm/mach-ixp4xx/common.c	2007-05-16 07:51:08.000000000 -0700
@@ -195,7 +195,7 @@
 
 static void ixp4xx_irq_mask(unsigned int irq)
 {
-	if ((cpu_is_ixp46x() || cpu_is_ixp43x()) && irq >= 32)
+	if (cpu_is_ixp46x() && irq >= 32)
 		*IXP4XX_ICMR2 &= ~(1 << (irq - 32));
 	else
 		*IXP4XX_ICMR &= ~(1 << irq);
@@ -218,7 +218,7 @@
 	if (!(ixp4xx_irq_edge & (1 << irq)))
 		ixp4xx_irq_ack(irq);
 
-	if ((cpu_is_ixp46x() || cpu_is_ixp43x()) && irq >= 32)
+	if (cpu_is_ixp46x() && irq >= 32)
 		*IXP4XX_ICMR2 |= (1 << (irq - 32));
 	else
 		*IXP4XX_ICMR |= (1 << irq);
@@ -242,7 +242,7 @@
 	/* Disable all interrupt */
 	*IXP4XX_ICMR = 0x0; 
 
-	if (cpu_is_ixp46x() || cpu_is_ixp43x()) {
+	if (cpu_is_ixp46x()) {
 		/* Route upper 32 sources to IRQ instead of FIQ */
 		*IXP4XX_ICLR2 = 0x00;
 
Index: linux-2.6.22-rc1-armeb/include/asm-arm/arch-ixp4xx/io.h
===================================================================
--- linux-2.6.22-rc1-armeb.orig/include/asm-arm/arch-ixp4xx/io.h	2007-05-16 05:55:08.000000000 -0700
+++ linux-2.6.22-rc1-armeb/include/asm-arm/arch-ixp4xx/io.h	2007-05-16 07:53:18.000000000 -0700
@@ -61,7 +61,7 @@
 static inline void __iomem *
 __ixp4xx_ioremap(unsigned long addr, size_t size, unsigned int mtype)
 {
-	if((addr < PCIBIOS_MIN_MEM) || (addr > 0x4fffffff))
+	if((addr < 0x48000000) || (addr > 0x4fffffff))
 		return __arm_ioremap(addr, size, mtype);
 
 	return (void *)addr;
Index: linux-2.6.22-rc1-armeb/arch/arm/configs/ixp4xx_defconfig
===================================================================
--- linux-2.6.22-rc1-armeb.orig/arch/arm/configs/ixp4xx_defconfig	2007-05-16 07:49:12.000000000 -0700
+++ linux-2.6.22-rc1-armeb/arch/arm/configs/ixp4xx_defconfig	2007-05-16 07:51:08.000000000 -0700
@@ -117,13 +117,11 @@
 CONFIG_ARCH_IXDP425=y
 CONFIG_MACH_IXDPG425=y
 CONFIG_MACH_IXDP465=y
-CONFIG_MACH_KIXRP435=y
 CONFIG_ARCH_IXCDP1100=y
 CONFIG_ARCH_PRPMC1100=y
 CONFIG_MACH_NAS100D=y
 CONFIG_ARCH_IXDP4XX=y
 CONFIG_CPU_IXP46X=y
-CONFIG_CPU_IXP43X=y
 # CONFIG_MACH_GTWX5715 is not set
 
 #
Index: linux-2.6.22-rc1-armeb/arch/arm/mach-ixp4xx/ixdp425-pci.c
===================================================================
--- linux-2.6.22-rc1-armeb.orig/arch/arm/mach-ixp4xx/ixdp425-pci.c	2007-05-16 05:54:59.000000000 -0700
+++ linux-2.6.22-rc1-armeb/arch/arm/mach-ixp4xx/ixdp425-pci.c	2007-05-16 07:51:08.000000000 -0700
@@ -66,7 +66,7 @@
 int __init ixdp425_pci_init(void)
 {
 	if (machine_is_ixdp425() || machine_is_ixcdp1100() ||
-			machine_is_ixdp465() || machine_is_kixrp435())
+			machine_is_ixdp465())
 		pci_common_init(&ixdp425_pci);
 	return 0;
 }
Index: linux-2.6.22-rc1-armeb/arch/arm/mach-ixp4xx/Kconfig
===================================================================
--- linux-2.6.22-rc1-armeb.orig/arch/arm/mach-ixp4xx/Kconfig	2007-05-16 07:48:53.000000000 -0700
+++ linux-2.6.22-rc1-armeb/arch/arm/mach-ixp4xx/Kconfig	2007-05-16 07:51:08.000000000 -0700
@@ -62,12 +62,6 @@
 	  IXDP465 Development Platform (Also known as BMP).
 	  For more information on this platform, see <file:Documentation/arm/IXP4xx>.
 
-config MACH_KIXRP435
-	bool "KIXRP435"
-	help
-	  Say 'Y' here if you want your kernel to support Intel's
-	  KIXRP435 Reference Platform.
-	  For more information on this platform, see <file:Documentation/arm/IXP4xx>.
 
 #
 # IXCDP1100 is the exact same HW as IXDP425, but with a different machine 
@@ -109,7 +103,7 @@
 #
 config	ARCH_IXDP4XX
 	bool
-	depends on ARCH_IXDP425 || MACH_IXDP465 || MACH_KIXRP435
+	depends on ARCH_IXDP425 || MACH_IXDP465
 	default y
 
 #
@@ -120,11 +114,6 @@
 	depends on MACH_IXDP465
 	default y
 
-config CPU_IXP43X
-	bool
-	depends on MACH_KIXRP435
-	default y
-
 config MACH_GTWX5715
 	bool "Gemtek WX5715 (Linksys WRV54G)"
 	depends on ARCH_IXP4XX
Index: linux-2.6.22-rc1-armeb/include/asm-arm/arch-ixp4xx/entry-macro.S
===================================================================
--- linux-2.6.22-rc1-armeb.orig/include/asm-arm/arch-ixp4xx/entry-macro.S	2007-05-16 05:55:08.000000000 -0700
+++ linux-2.6.22-rc1-armeb/include/asm-arm/arch-ixp4xx/entry-macro.S	2007-05-16 07:51:08.000000000 -0700
@@ -31,9 +31,9 @@
 
 1001:
 		/*
-		 * IXP465/IXP435 has an upper IRQ status register
+		 * IXP465 has an upper IRQ status register
 		 */
-#if defined(CONFIG_CPU_IXP46X) || defined(CONFIG_CPU_IXP43X)
+#if defined(CONFIG_CPU_IXP46X)
 		ldr	\irqstat, =(IXP4XX_INTC_BASE_VIRT+IXP4XX_ICIP2_OFFSET)
 		ldr	\irqstat, [\irqstat]		@ get upper interrupts
 		mov	\irqnr, #63
Index: linux-2.6.22-rc1-armeb/arch/arm/mach-ixp4xx/ixdp425-setup.c
===================================================================
--- linux-2.6.22-rc1-armeb.orig/arch/arm/mach-ixp4xx/ixdp425-setup.c	2007-05-16 07:48:53.000000000 -0700
+++ linux-2.6.22-rc1-armeb/arch/arm/mach-ixp4xx/ixdp425-setup.c	2007-05-16 07:51:08.000000000 -0700
@@ -115,11 +115,6 @@
 	ixdp425_flash_resource.end =
 		IXP4XX_EXP_BUS_BASE(0) + ixp4xx_exp_bus_size - 1;
 
-	if (cpu_is_ixp43x()) {
-		ixdp425_uart.num_resources = 1;
-		ixdp425_uart_data[1].flags = 0;
-	}
-
 	platform_add_devices(ixdp425_devices, ARRAY_SIZE(ixdp425_devices));
 }
 
@@ -161,16 +156,3 @@
 	.init_machine	= ixdp425_init,
 MACHINE_END
 #endif
-
-#ifdef CONFIG_MACH_KIXRP435
-MACHINE_START(KIXRP435, "Intel KIXRP435 Reference Platform")
-	/* Maintainer: MontaVista Software, Inc. */
-	.phys_io	= IXP4XX_PERIPHERAL_BASE_PHYS,
-	.io_pg_offst	= ((IXP4XX_PERIPHERAL_BASE_VIRT) >> 18) & 0xfffc,
-	.map_io		= ixp4xx_map_io,
-	.init_irq	= ixp4xx_init_irq,
-	.timer		= &ixp4xx_timer,
-	.boot_params	= 0x0100,
-	.init_machine	= ixdp425_init,
-MACHINE_END
-#endif
