Index: linux-2.6.22-rc3-git2-armeb/arch/arm/mach-ixp4xx/gtwx5715-pci.c
===================================================================
--- linux-2.6.22-rc3-git2-armeb.orig/arch/arm/mach-ixp4xx/gtwx5715-pci.c	2007-05-31 04:06:05.000000000 -0700
+++ linux-2.6.22-rc3-git2-armeb/arch/arm/mach-ixp4xx/gtwx5715-pci.c	2007-05-31 04:12:02.000000000 -0700
@@ -25,17 +25,14 @@
 #include <linux/pci.h>
 #include <linux/init.h>
 #include <linux/delay.h>
+#include <linux/irq.h>
+
 #include <asm/mach-types.h>
 #include <asm/hardware.h>
 #include <asm/irq.h>
 #include <asm/arch/gtwx5715.h>
 #include <asm/mach/pci.h>
 
-extern void ixp4xx_pci_preinit(void);
-extern int ixp4xx_setup(int nr, struct pci_sys_data *sys);
-extern struct pci_bus *ixp4xx_scan_bus(int nr, struct pci_sys_data *sys);
-
-
 /*
  * The exact GPIO pins and IRQs are defined in arch-ixp4xx/gtwx5715.h
  * Slot 0 isn't actually populated with a card connector but
