Index: linux-2.6.18/drivers/net/wireless/zd1211rw/zd_usb.c
===================================================================
--- linux-2.6.18.orig/drivers/net/wireless/zd1211rw/zd_usb.c	2006-10-02 00:51:44.000000000 +0200
+++ linux-2.6.18/drivers/net/wireless/zd1211rw/zd_usb.c	2006-10-02 00:51:46.000000000 +0200
@@ -574,7 +574,7 @@
 {
 	int i;
 	struct zd_mac *mac = zd_usb_to_mac(usb);
-	const struct rx_length_info *length_info;
+	struct rx_length_info *length_info;
 
 	if (length < sizeof(struct rx_length_info)) {
 		/* It's not a complete packet anyhow. */
