Signed-off-by: Leon Woestenberg <leonw@mailcan.com>

Index: linux-2.6.20-rc1-git5/fs/Makefile
===================================================================
--- linux-2.6.20-rc1-git5.orig/fs/Makefile	2006-12-19 22:37:44.175903713 +0000
+++ linux-2.6.20-rc1-git5/fs/Makefile	2006-12-19 22:38:11.397604965 +0000
@@ -68,6 +68,7 @@
 obj-$(CONFIG_JBD2)		+= jbd2/
 obj-$(CONFIG_EXT2_FS)		+= ext2/
 obj-$(CONFIG_CRAMFS)		+= cramfs/
+obj-$(CONFIG_SQUASHFS)		+= squashfs/
 obj-$(CONFIG_RAMFS)		+= ramfs/
 obj-$(CONFIG_HUGETLBFS)		+= hugetlbfs/
 obj-$(CONFIG_CODA_FS)		+= coda/
