Index: linux-2.6.19/net/bridge/br_input.c
===================================================================
--- linux-2.6.19.orig/net/bridge/br_input.c
+++ linux-2.6.19/net/bridge/br_input.c
@@ -32,6 +32,9 @@ static void br_pass_frame_up(struct net_
 	indev = skb->dev;
 	skb->dev = br->dev;
 
+	skb_push(skb, ETH_HLEN);
+	skb->protocol = eth_type_trans(skb, skb->dev);
+
 	NF_HOOK(PF_BRIDGE, NF_BR_LOCAL_IN, skb, indev, NULL,
 		netif_receive_skb);
 }
