Index: linux-2.6.17/arch/arm/common/dmabounce.c
===================================================================
--- linux-2.6.17.orig/arch/arm/common/dmabounce.c	2006-09-05 00:51:14.000000000 +0200
+++ linux-2.6.17/arch/arm/common/dmabounce.c	2006-09-05 02:20:46.000000000 +0200
@@ -663,6 +663,7 @@
 EXPORT_SYMBOL(dma_map_sg);
 EXPORT_SYMBOL(dma_unmap_sg);
 EXPORT_SYMBOL(dma_sync_single);
+EXPORT_SYMBOL(dma_sync_single_for_device);
 EXPORT_SYMBOL(dma_sync_sg);
 EXPORT_SYMBOL(dmabounce_register_dev);
 EXPORT_SYMBOL(dmabounce_unregister_dev);
