Index: linux-2.6.21-rc4-arm/arch/arm/mach-ixp4xx/nslu2-setup.c
===================================================================
--- linux-2.6.21-rc4-arm.orig/arch/arm/mach-ixp4xx/nslu2-setup.c	2007-03-20 01:30:47.000000000 -0800
+++ linux-2.6.21-rc4-arm/arch/arm/mach-ixp4xx/nslu2-setup.c	2007-03-20 01:39:29.000000000 -0800
@@ -20,6 +20,7 @@
 #include <linux/leds.h>
 #include <linux/mtd/mtd.h>
 
+#include <asm/setup.h>
 #include <asm/mach-types.h>
 #include <asm/mach/arch.h>
 #include <asm/mach/flash.h>
@@ -238,10 +239,36 @@
 	register_mtd_user(&nslu2_flash_notifier);
 }
 
-static void __init nslu2_fixup(void)
+static char nslu2_rtc_probe[] __initdata = "rtc-x1205.probe=0,0x6f ";
+        
+static void __init nslu2_fixup(struct machine_desc *desc,
+                struct tag *tags, char **cmdline, struct meminfo *mi)
 {
+        struct tag *t = tags;
+        char *p = *cmdline;
+
 	/* The xtal on this machine is non-standard. */
 	ixp4xx_timer_freq = NSLU2_FREQ;
+
+        /* Find the end of the tags table, taking note of any cmdline tag. */
+        for (; t->hdr.size; t = tag_next(t)) {
+                if (t->hdr.tag == ATAG_CMDLINE) {
+                        p = t->u.cmdline.cmdline;
+                }
+        }
+
+        /* Overwrite the end of the table with a new cmdline tag. */
+        t->hdr.tag = ATAG_CMDLINE;
+        t->hdr.size = (sizeof (struct tag_header) +
+                       strlen(nslu2_rtc_probe) + strlen(p) + 1 + 4) >> 2;
+        strlcpy(t->u.cmdline.cmdline, nslu2_rtc_probe, COMMAND_LINE_SIZE);
+        strlcpy(t->u.cmdline.cmdline + strlen(nslu2_rtc_probe), p,
+                COMMAND_LINE_SIZE - strlen(nslu2_rtc_probe));
+
+        /* Terminate the table. */
+        t = tag_next(t);
+        t->hdr.tag = ATAG_NONE;
+        t->hdr.size = 0;
 }
 
 MACHINE_START(NSLU2, "Linksys NSLU2")
