Index: apex-1.5.8/src/mach-ixp42x/debian-nas100d-arm_config
===================================================================
--- apex-1.5.8.orig/src/mach-ixp42x/debian-nas100d-arm_config
+++ apex-1.5.8/src/mach-ixp42x/debian-nas100d-arm_config
@@ -19,7 +19,7 @@ CONFIG_EXPERIMENTAL=y
 #
 CONFIG_TARGET_DESCRIPTION="NAS100D"
 CONFIG_CROSS_COMPILE=""
-# CONFIG_AEABI is not set
+CONFIG_AEABI=y
 CONFIG_CC_OPTIMIZE_FOR_SIZE=y
 # CONFIG_CC_OPTIMIZE_FOR_SPEED is not set
 # CONFIG_CC_NO_OPTIMIZATION is not set
Index: apex-1.5.8/src/mach-ixp42x/debian-nas100d-armeb_config
===================================================================
--- apex-1.5.8.orig/src/mach-ixp42x/debian-nas100d-armeb_config
+++ apex-1.5.8/src/mach-ixp42x/debian-nas100d-armeb_config
@@ -19,7 +19,7 @@ CONFIG_EXPERIMENTAL=y
 #
 CONFIG_TARGET_DESCRIPTION="NAS100D"
 CONFIG_CROSS_COMPILE=""
-# CONFIG_AEABI is not set
+CONFIG_AEABI=y
 CONFIG_CC_OPTIMIZE_FOR_SIZE=y
 # CONFIG_CC_OPTIMIZE_FOR_SPEED is not set
 # CONFIG_CC_NO_OPTIMIZATION is not set
Index: apex-1.5.8/src/mach-ixp42x/debian-nslu2-16mb-arm_config
===================================================================
--- apex-1.5.8.orig/src/mach-ixp42x/debian-nslu2-16mb-arm_config
+++ apex-1.5.8/src/mach-ixp42x/debian-nslu2-16mb-arm_config
@@ -19,7 +19,7 @@ CONFIG_EXPERIMENTAL=y
 #
 CONFIG_TARGET_DESCRIPTION="Debian NSLU2/LE (16MiB Flash)"
 CONFIG_CROSS_COMPILE=""
-# CONFIG_AEABI is not set
+CONFIG_AEABI=y
 CONFIG_CC_OPTIMIZE_FOR_SIZE=y
 # CONFIG_CC_OPTIMIZE_FOR_SPEED is not set
 # CONFIG_CC_NO_OPTIMIZATION is not set
Index: apex-1.5.8/src/mach-ixp42x/debian-nslu2-16mb-armeb_config
===================================================================
--- apex-1.5.8.orig/src/mach-ixp42x/debian-nslu2-16mb-armeb_config
+++ apex-1.5.8/src/mach-ixp42x/debian-nslu2-16mb-armeb_config
@@ -19,7 +19,7 @@ CONFIG_EXPERIMENTAL=y
 #
 CONFIG_TARGET_DESCRIPTION="Debian NSLU2/BE (16MiB Flash)"
 CONFIG_CROSS_COMPILE=""
-# CONFIG_AEABI is not set
+CONFIG_AEABI=y
 CONFIG_CC_OPTIMIZE_FOR_SIZE=y
 # CONFIG_CC_OPTIMIZE_FOR_SPEED is not set
 # CONFIG_CC_NO_OPTIMIZATION is not set
Index: apex-1.5.8/src/mach-ixp42x/debian-nslu2-arm_config
===================================================================
--- apex-1.5.8.orig/src/mach-ixp42x/debian-nslu2-arm_config
+++ apex-1.5.8/src/mach-ixp42x/debian-nslu2-arm_config
@@ -19,7 +19,7 @@ CONFIG_EXPERIMENTAL=y
 #
 CONFIG_TARGET_DESCRIPTION="Debian NSLU2"
 CONFIG_CROSS_COMPILE=""
-# CONFIG_AEABI is not set
+CONFIG_AEABI=y
 CONFIG_CC_OPTIMIZE_FOR_SIZE=y
 # CONFIG_CC_OPTIMIZE_FOR_SPEED is not set
 # CONFIG_CC_NO_OPTIMIZATION is not set
Index: apex-1.5.8/src/mach-ixp42x/debian-nslu2-armeb_config
===================================================================
--- apex-1.5.8.orig/src/mach-ixp42x/debian-nslu2-armeb_config
+++ apex-1.5.8/src/mach-ixp42x/debian-nslu2-armeb_config
@@ -19,7 +19,7 @@ CONFIG_EXPERIMENTAL=y
 #
 CONFIG_TARGET_DESCRIPTION="Debian NSLU2 (bigendian)"
 CONFIG_CROSS_COMPILE=""
-# CONFIG_AEABI is not set
+CONFIG_AEABI=y
 CONFIG_CC_OPTIMIZE_FOR_SIZE=y
 # CONFIG_CC_OPTIMIZE_FOR_SPEED is not set
 # CONFIG_CC_NO_OPTIMIZATION is not set
Index: apex-1.5.8/src/mach-ixp42x/nslu2_config
===================================================================
--- apex-1.5.8.orig/src/mach-ixp42x/nslu2_config
+++ apex-1.5.8/src/mach-ixp42x/nslu2_config
@@ -19,7 +19,7 @@ CONFIG_ARCH_IXP42X=y
 #
 CONFIG_TARGET_DESCRIPTION="NSLU2"
 CONFIG_CROSS_COMPILE="/usr/armv5b-softfloat-linux/gcc3/armv5b-softfloat-linux-"
-# CONFIG_AEABI is not set
+CONFIG_AEABI=y
 CONFIG_CC_OPTIMIZE_FOR_SIZE=y
 # CONFIG_CC_OPTIMIZE_FOR_SPEED is not set
 # CONFIG_CC_NO_OPTIMIZATION is not set
Index: apex-1.5.8/src/mach-ixp42x/slugos-dsmg600-arm_config
===================================================================
--- apex-1.5.8.orig/src/mach-ixp42x/slugos-dsmg600-arm_config
+++ apex-1.5.8/src/mach-ixp42x/slugos-dsmg600-arm_config
@@ -19,7 +19,7 @@ CONFIG_EXPERIMENTAL=y
 #
 CONFIG_TARGET_DESCRIPTION="SlugOS DMSG600/LE"
 CONFIG_CROSS_COMPILE=""
-# CONFIG_AEABI is not set
+CONFIG_AEABI=y
 CONFIG_CC_OPTIMIZE_FOR_SIZE=y
 # CONFIG_CC_OPTIMIZE_FOR_SPEED is not set
 # CONFIG_CC_NO_OPTIMIZATION is not set
Index: apex-1.5.8/src/mach-ixp42x/slugos-dsmg600-armeb_config
===================================================================
--- apex-1.5.8.orig/src/mach-ixp42x/slugos-dsmg600-armeb_config
+++ apex-1.5.8/src/mach-ixp42x/slugos-dsmg600-armeb_config
@@ -19,7 +19,7 @@ CONFIG_EXPERIMENTAL=y
 #
 CONFIG_TARGET_DESCRIPTION="SlugOS DMSG600/BE"
 CONFIG_CROSS_COMPILE=""
-# CONFIG_AEABI is not set
+CONFIG_AEABI=y
 CONFIG_CC_OPTIMIZE_FOR_SIZE=y
 # CONFIG_CC_OPTIMIZE_FOR_SPEED is not set
 # CONFIG_CC_NO_OPTIMIZATION is not set
Index: apex-1.5.8/src/mach-ixp42x/slugos-nas100d-arm_config
===================================================================
--- apex-1.5.8.orig/src/mach-ixp42x/slugos-nas100d-arm_config
+++ apex-1.5.8/src/mach-ixp42x/slugos-nas100d-arm_config
@@ -19,7 +19,7 @@ CONFIG_EXPERIMENTAL=y
 #
 CONFIG_TARGET_DESCRIPTION="SlugOS NAS100D"
 CONFIG_CROSS_COMPILE=""
-# CONFIG_AEABI is not set
+CONFIG_AEABI=y
 CONFIG_CC_OPTIMIZE_FOR_SIZE=y
 # CONFIG_CC_OPTIMIZE_FOR_SPEED is not set
 # CONFIG_CC_NO_OPTIMIZATION is not set
Index: apex-1.5.8/src/mach-ixp42x/slugos-nas100d-armeb_config
===================================================================
--- apex-1.5.8.orig/src/mach-ixp42x/slugos-nas100d-armeb_config
+++ apex-1.5.8/src/mach-ixp42x/slugos-nas100d-armeb_config
@@ -19,7 +19,7 @@ CONFIG_EXPERIMENTAL=y
 #
 CONFIG_TARGET_DESCRIPTION="SlugOS NAS100D (bigendian)"
 CONFIG_CROSS_COMPILE=""
-# CONFIG_AEABI is not set
+CONFIG_AEABI=y
 CONFIG_CC_OPTIMIZE_FOR_SIZE=y
 # CONFIG_CC_OPTIMIZE_FOR_SPEED is not set
 # CONFIG_CC_NO_OPTIMIZATION is not set
Index: apex-1.5.8/src/mach-ixp42x/slugos-nslu2-16mb-arm_config
===================================================================
--- apex-1.5.8.orig/src/mach-ixp42x/slugos-nslu2-16mb-arm_config
+++ apex-1.5.8/src/mach-ixp42x/slugos-nslu2-16mb-arm_config
@@ -19,7 +19,7 @@ CONFIG_EXPERIMENTAL=y
 #
 CONFIG_TARGET_DESCRIPTION="SlugOS NSLU2/LE (16MiB Flash)"
 CONFIG_CROSS_COMPILE=""
-# CONFIG_AEABI is not set
+CONFIG_AEABI=y
 CONFIG_CC_OPTIMIZE_FOR_SIZE=y
 # CONFIG_CC_OPTIMIZE_FOR_SPEED is not set
 # CONFIG_CC_NO_OPTIMIZATION is not set
Index: apex-1.5.8/src/mach-ixp42x/slugos-nslu2-16mb-armeb_config
===================================================================
--- apex-1.5.8.orig/src/mach-ixp42x/slugos-nslu2-16mb-armeb_config
+++ apex-1.5.8/src/mach-ixp42x/slugos-nslu2-16mb-armeb_config
@@ -19,7 +19,7 @@ CONFIG_EXPERIMENTAL=y
 #
 CONFIG_TARGET_DESCRIPTION="SlugOS NSLU2/BE (16MiB Flash)"
 CONFIG_CROSS_COMPILE=""
-# CONFIG_AEABI is not set
+CONFIG_AEABI=y
 CONFIG_CC_OPTIMIZE_FOR_SIZE=y
 # CONFIG_CC_OPTIMIZE_FOR_SPEED is not set
 # CONFIG_CC_NO_OPTIMIZATION is not set
Index: apex-1.5.8/src/mach-ixp42x/slugos-nslu2-arm_config
===================================================================
--- apex-1.5.8.orig/src/mach-ixp42x/slugos-nslu2-arm_config
+++ apex-1.5.8/src/mach-ixp42x/slugos-nslu2-arm_config
@@ -19,7 +19,7 @@ CONFIG_EXPERIMENTAL=y
 #
 CONFIG_TARGET_DESCRIPTION="SlugOS NSLU2"
 CONFIG_CROSS_COMPILE=""
-# CONFIG_AEABI is not set
+CONFIG_AEABI=y
 CONFIG_CC_OPTIMIZE_FOR_SIZE=y
 # CONFIG_CC_OPTIMIZE_FOR_SPEED is not set
 # CONFIG_CC_NO_OPTIMIZATION is not set
Index: apex-1.5.8/src/mach-ixp42x/slugos-nslu2-armeb_config
===================================================================
--- apex-1.5.8.orig/src/mach-ixp42x/slugos-nslu2-armeb_config
+++ apex-1.5.8/src/mach-ixp42x/slugos-nslu2-armeb_config
@@ -19,7 +19,7 @@ CONFIG_EXPERIMENTAL=y
 #
 CONFIG_TARGET_DESCRIPTION="SlugOS NSLU2 (bigendian)"
 CONFIG_CROSS_COMPILE=""
-# CONFIG_AEABI is not set
+CONFIG_AEABI=y
 CONFIG_CC_OPTIMIZE_FOR_SIZE=y
 # CONFIG_CC_OPTIMIZE_FOR_SPEED is not set
 # CONFIG_CC_NO_OPTIMIZATION is not set
