Index: apex-1.5.13/src/mach-ixp42x/debian-dsmg600-arm_config
===================================================================
--- apex-1.5.13.orig/src/mach-ixp42x/debian-dsmg600-arm_config	2008-01-27 21:11:09.000000000 +1030
+++ apex-1.5.13/src/mach-ixp42x/debian-dsmg600-arm_config	2008-01-27 21:18:28.000000000 +1030
@@ -140,7 +140,8 @@
 # CONFIG_ENV_REGION_KERNEL_SWAP is not set
 CONFIG_ENV_STARTUP_RAMDISK_COPY=y
 # CONFIG_ENV_REGION_RAMDISK_SWAP is not set
-# CONFIG_ENV_STARTUP_PREFIX_P is not set
+CONFIG_ENV_STARTUP_PREFIX_P=y
+CONFIG_ENV_STARTUP_PREFIX="sdram-init; memscan -u 0+256m"
 
 #
 #    Regions
Index: apex-1.5.13/src/mach-ixp42x/debian-dsmg600-armeb_config
===================================================================
--- apex-1.5.13.orig/src/mach-ixp42x/debian-dsmg600-armeb_config	2008-01-27 21:11:09.000000000 +1030
+++ apex-1.5.13/src/mach-ixp42x/debian-dsmg600-armeb_config	2008-01-27 21:18:35.000000000 +1030
@@ -140,7 +140,8 @@
 # CONFIG_ENV_REGION_KERNEL_SWAP is not set
 CONFIG_ENV_STARTUP_RAMDISK_COPY=y
 # CONFIG_ENV_REGION_RAMDISK_SWAP is not set
-# CONFIG_ENV_STARTUP_PREFIX_P is not set
+CONFIG_ENV_STARTUP_PREFIX_P=y
+CONFIG_ENV_STARTUP_PREFIX="sdram-init; memscan -u 0+256m"
 
 #
 #    Regions
Index: apex-1.5.13/src/mach-ixp42x/debian-nas100d-arm_config
===================================================================
--- apex-1.5.13.orig/src/mach-ixp42x/debian-nas100d-arm_config	2008-01-27 21:11:10.000000000 +1030
+++ apex-1.5.13/src/mach-ixp42x/debian-nas100d-arm_config	2008-01-27 21:19:14.000000000 +1030
@@ -137,7 +137,8 @@
 # CONFIG_ENV_REGION_KERNEL_SWAP is not set
 CONFIG_ENV_STARTUP_RAMDISK_COPY=y
 # CONFIG_ENV_REGION_RAMDISK_SWAP is not set
-# CONFIG_ENV_STARTUP_PREFIX_P is not set
+CONFIG_ENV_STARTUP_PREFIX_P=y
+CONFIG_ENV_STARTUP_PREFIX="sdram-init; memscan -u 0+256m"
 
 #
 #    Regions
Index: apex-1.5.13/src/mach-ixp42x/debian-nas100d-armeb_config
===================================================================
--- apex-1.5.13.orig/src/mach-ixp42x/debian-nas100d-armeb_config	2008-01-27 21:11:10.000000000 +1030
+++ apex-1.5.13/src/mach-ixp42x/debian-nas100d-armeb_config	2008-01-27 21:19:18.000000000 +1030
@@ -137,7 +137,8 @@
 # CONFIG_ENV_REGION_KERNEL_SWAP is not set
 CONFIG_ENV_STARTUP_RAMDISK_COPY=y
 # CONFIG_ENV_REGION_RAMDISK_SWAP is not set
-# CONFIG_ENV_STARTUP_PREFIX_P is not set
+CONFIG_ENV_STARTUP_PREFIX_P=y
+CONFIG_ENV_STARTUP_PREFIX="sdram-init; memscan -u 0+256m"
 
 #
 #    Regions
Index: apex-1.5.13/src/mach-ixp42x/debian-nslu2-16mb-arm_config
===================================================================
--- apex-1.5.13.orig/src/mach-ixp42x/debian-nslu2-16mb-arm_config	2008-01-27 21:11:10.000000000 +1030
+++ apex-1.5.13/src/mach-ixp42x/debian-nslu2-16mb-arm_config	2008-01-27 21:19:59.000000000 +1030
@@ -141,7 +141,7 @@
 CONFIG_ENV_STARTUP_RAMDISK_COPY=y
 CONFIG_ENV_REGION_RAMDISK_SWAP=y
 CONFIG_ENV_STARTUP_PREFIX_P=y
-CONFIG_ENV_STARTUP_PREFIX="sdram-init; memscan -u 0+64m"
+CONFIG_ENV_STARTUP_PREFIX="sdram-init; memscan -u 0+256m"
 
 #
 #    Regions
Index: apex-1.5.13/src/mach-ixp42x/debian-nslu2-16mb-armeb_config
===================================================================
--- apex-1.5.13.orig/src/mach-ixp42x/debian-nslu2-16mb-armeb_config	2008-01-27 21:11:10.000000000 +1030
+++ apex-1.5.13/src/mach-ixp42x/debian-nslu2-16mb-armeb_config	2008-01-27 21:19:44.000000000 +1030
@@ -141,7 +141,7 @@
 CONFIG_ENV_STARTUP_RAMDISK_COPY=y
 # CONFIG_ENV_REGION_RAMDISK_SWAP is not set
 CONFIG_ENV_STARTUP_PREFIX_P=y
-CONFIG_ENV_STARTUP_PREFIX="sdram-init; memscan -u 0+64m"
+CONFIG_ENV_STARTUP_PREFIX="sdram-init; memscan -u 0+256m"
 
 #
 #    Regions
Index: apex-1.5.13/src/mach-ixp42x/debian-nslu2-armeb_config
===================================================================
--- apex-1.5.13.orig/src/mach-ixp42x/debian-nslu2-armeb_config	2008-01-27 21:11:10.000000000 +1030
+++ apex-1.5.13/src/mach-ixp42x/debian-nslu2-armeb_config	2008-01-27 21:20:17.000000000 +1030
@@ -140,7 +140,8 @@
 # CONFIG_ENV_REGION_KERNEL_SWAP is not set
 CONFIG_ENV_STARTUP_RAMDISK_COPY=y
 # CONFIG_ENV_REGION_RAMDISK_SWAP is not set
-# CONFIG_ENV_STARTUP_PREFIX_P is not set
+CONFIG_ENV_STARTUP_PREFIX_P=y
+CONFIG_ENV_STARTUP_PREFIX="sdram-init; memscan -u 0+256m"
 
 #
 #    Regions
Index: apex-1.5.13/src/mach-ixp42x/slugos-dsmg600-arm_config
===================================================================
--- apex-1.5.13.orig/src/mach-ixp42x/slugos-dsmg600-arm_config	2008-01-27 21:11:10.000000000 +1030
+++ apex-1.5.13/src/mach-ixp42x/slugos-dsmg600-arm_config	2008-01-27 21:20:55.000000000 +1030
@@ -133,7 +133,8 @@
 CONFIG_AUTOBOOT_DELAY=10
 CONFIG_ENV_STARTUP_KERNEL_COPY=y
 # CONFIG_ENV_REGION_KERNEL_SWAP is not set
-# CONFIG_ENV_STARTUP_PREFIX_P is not set
+CONFIG_ENV_STARTUP_PREFIX_P=y
+CONFIG_ENV_STARTUP_PREFIX="sdram-init; memscan -u 0+256m"
 
 #
 #    Regions
Index: apex-1.5.13/src/mach-ixp42x/slugos-dsmg600-armeb_config
===================================================================
--- apex-1.5.13.orig/src/mach-ixp42x/slugos-dsmg600-armeb_config	2008-01-27 21:11:10.000000000 +1030
+++ apex-1.5.13/src/mach-ixp42x/slugos-dsmg600-armeb_config	2008-01-27 21:20:58.000000000 +1030
@@ -133,7 +133,8 @@
 CONFIG_AUTOBOOT_DELAY=10
 CONFIG_ENV_STARTUP_KERNEL_COPY=y
 # CONFIG_ENV_REGION_KERNEL_SWAP is not set
-# CONFIG_ENV_STARTUP_PREFIX_P is not set
+CONFIG_ENV_STARTUP_PREFIX_P=y
+CONFIG_ENV_STARTUP_PREFIX="sdram-init; memscan -u 0+256m"
 
 #
 #    Regions
Index: apex-1.5.13/src/mach-ixp42x/slugos-fsg3-arm_config
===================================================================
--- apex-1.5.13.orig/src/mach-ixp42x/slugos-fsg3-arm_config	2008-01-27 21:11:10.000000000 +1030
+++ apex-1.5.13/src/mach-ixp42x/slugos-fsg3-arm_config	2008-01-27 21:21:05.000000000 +1030
@@ -134,7 +134,8 @@
 CONFIG_AUTOBOOT_DELAY=10
 CONFIG_ENV_STARTUP_KERNEL_COPY=y
 # CONFIG_ENV_REGION_KERNEL_SWAP is not set
-# CONFIG_ENV_STARTUP_PREFIX_P is not set
+CONFIG_ENV_STARTUP_PREFIX_P=y
+CONFIG_ENV_STARTUP_PREFIX="sdram-init; memscan -u 0+256m"
 
 #
 #    Regions
Index: apex-1.5.13/src/mach-ixp42x/slugos-fsg3-armeb_config
===================================================================
--- apex-1.5.13.orig/src/mach-ixp42x/slugos-fsg3-armeb_config	2008-01-27 21:11:10.000000000 +1030
+++ apex-1.5.13/src/mach-ixp42x/slugos-fsg3-armeb_config	2008-01-27 21:21:02.000000000 +1030
@@ -134,7 +134,8 @@
 CONFIG_AUTOBOOT_DELAY=10
 CONFIG_ENV_STARTUP_KERNEL_COPY=y
 # CONFIG_ENV_REGION_KERNEL_SWAP is not set
-# CONFIG_ENV_STARTUP_PREFIX_P is not set
+CONFIG_ENV_STARTUP_PREFIX_P=y
+CONFIG_ENV_STARTUP_PREFIX="sdram-init; memscan -u 0+256m"
 
 #
 #    Regions
Index: apex-1.5.13/src/mach-ixp42x/slugos-nas100d-arm_config
===================================================================
--- apex-1.5.13.orig/src/mach-ixp42x/slugos-nas100d-arm_config	2008-01-27 21:11:10.000000000 +1030
+++ apex-1.5.13/src/mach-ixp42x/slugos-nas100d-arm_config	2008-01-27 21:21:15.000000000 +1030
@@ -133,7 +133,8 @@
 CONFIG_AUTOBOOT_DELAY=10
 CONFIG_ENV_STARTUP_KERNEL_COPY=y
 # CONFIG_ENV_REGION_KERNEL_SWAP is not set
-# CONFIG_ENV_STARTUP_PREFIX_P is not set
+CONFIG_ENV_STARTUP_PREFIX_P=y
+CONFIG_ENV_STARTUP_PREFIX="sdram-init; memscan -u 0+256m"
 
 #
 #    Regions
Index: apex-1.5.13/src/mach-ixp42x/slugos-nas100d-armeb_config
===================================================================
--- apex-1.5.13.orig/src/mach-ixp42x/slugos-nas100d-armeb_config	2008-01-27 21:11:10.000000000 +1030
+++ apex-1.5.13/src/mach-ixp42x/slugos-nas100d-armeb_config	2008-01-27 21:21:18.000000000 +1030
@@ -133,7 +133,8 @@
 CONFIG_AUTOBOOT_DELAY=10
 CONFIG_ENV_STARTUP_KERNEL_COPY=y
 # CONFIG_ENV_REGION_KERNEL_SWAP is not set
-# CONFIG_ENV_STARTUP_PREFIX_P is not set
+CONFIG_ENV_STARTUP_PREFIX_P=y
+CONFIG_ENV_STARTUP_PREFIX="sdram-init; memscan -u 0+256m"
 
 #
 #    Regions
Index: apex-1.5.13/src/mach-ixp42x/slugos-nslu2-16mb-arm_config
===================================================================
--- apex-1.5.13.orig/src/mach-ixp42x/slugos-nslu2-16mb-arm_config	2008-01-27 21:11:10.000000000 +1030
+++ apex-1.5.13/src/mach-ixp42x/slugos-nslu2-16mb-arm_config	2008-01-27 21:21:28.000000000 +1030
@@ -136,7 +136,8 @@
 CONFIG_AUTOBOOT_DELAY=10
 CONFIG_ENV_STARTUP_KERNEL_COPY=y
 # CONFIG_ENV_REGION_KERNEL_SWAP is not set
-# CONFIG_ENV_STARTUP_PREFIX_P is not set
+CONFIG_ENV_STARTUP_PREFIX_P=y
+CONFIG_ENV_STARTUP_PREFIX="sdram-init; memscan -u 0+256m"
 
 #
 #    Regions
Index: apex-1.5.13/src/mach-ixp42x/slugos-nslu2-16mb-armeb_config
===================================================================
--- apex-1.5.13.orig/src/mach-ixp42x/slugos-nslu2-16mb-armeb_config	2008-01-27 21:11:10.000000000 +1030
+++ apex-1.5.13/src/mach-ixp42x/slugos-nslu2-16mb-armeb_config	2008-01-27 21:21:31.000000000 +1030
@@ -136,7 +136,8 @@
 CONFIG_AUTOBOOT_DELAY=10
 CONFIG_ENV_STARTUP_KERNEL_COPY=y
 # CONFIG_ENV_REGION_KERNEL_SWAP is not set
-# CONFIG_ENV_STARTUP_PREFIX_P is not set
+CONFIG_ENV_STARTUP_PREFIX_P=y
+CONFIG_ENV_STARTUP_PREFIX="sdram-init; memscan -u 0+256m"
 
 #
 #    Regions
Index: apex-1.5.13/src/mach-ixp42x/slugos-nslu2-arm_config
===================================================================
--- apex-1.5.13.orig/src/mach-ixp42x/slugos-nslu2-arm_config	2008-01-27 21:11:10.000000000 +1030
+++ apex-1.5.13/src/mach-ixp42x/slugos-nslu2-arm_config	2008-01-27 21:21:25.000000000 +1030
@@ -136,7 +136,8 @@
 CONFIG_AUTOBOOT_DELAY=10
 CONFIG_ENV_STARTUP_KERNEL_COPY=y
 # CONFIG_ENV_REGION_KERNEL_SWAP is not set
-# CONFIG_ENV_STARTUP_PREFIX_P is not set
+CONFIG_ENV_STARTUP_PREFIX_P=y
+CONFIG_ENV_STARTUP_PREFIX="sdram-init; memscan -u 0+256m"
 
 #
 #    Regions
Index: apex-1.5.13/src/mach-ixp42x/slugos-nslu2-armeb_config
===================================================================
--- apex-1.5.13.orig/src/mach-ixp42x/slugos-nslu2-armeb_config	2008-01-27 21:11:10.000000000 +1030
+++ apex-1.5.13/src/mach-ixp42x/slugos-nslu2-armeb_config	2008-01-27 21:21:23.000000000 +1030
@@ -136,7 +136,8 @@
 CONFIG_AUTOBOOT_DELAY=10
 CONFIG_ENV_STARTUP_KERNEL_COPY=y
 # CONFIG_ENV_REGION_KERNEL_SWAP is not set
-# CONFIG_ENV_STARTUP_PREFIX_P is not set
+CONFIG_ENV_STARTUP_PREFIX_P=y
+CONFIG_ENV_STARTUP_PREFIX="sdram-init; memscan -u 0+256m"
 
 #
 #    Regions
