---
 src/mach-ixp42x/initialize-sdram.c |    1 +
 1 file changed, 1 insertion(+)

Index: apex-1.4.17/src/mach-ixp42x/initialize-sdram.c
===================================================================
--- apex-1.4.17.orig/src/mach-ixp42x/initialize-sdram.c
+++ apex-1.4.17/src/mach-ixp42x/initialize-sdram.c
@@ -51,6 +51,7 @@
 #include "hardware.h"
 #include <linux/kernel.h>
 #include <asm/mmu.h>
+#include <asm/cp15.h>
 #include "sdram.h"
 #include <drv-mem.h>
 #include <linux/string.h>
