Index: apex-1.5.13/src/mach-ixp42x/debian-nas100d-arm_config
===================================================================
--- apex-1.5.13.orig/src/mach-ixp42x/debian-nas100d-arm_config	2007-12-17 22:09:37.000000000 +1030
+++ apex-1.5.13/src/mach-ixp42x/debian-nas100d-arm_config	2007-12-17 22:11:34.000000000 +1030
@@ -17,7 +17,7 @@
 #
 # General Setup
 #
-CONFIG_TARGET_DESCRIPTION="NAS100D"
+CONFIG_TARGET_DESCRIPTION="Debian NAS100D/LE"
 CONFIG_CROSS_COMPILE=""
 CONFIG_AEABI=y
 CONFIG_CC_OPTIMIZE_FOR_SIZE=y
@@ -92,7 +92,7 @@
 CONFIG_CMD_XRECEIVE=y
 CONFIG_CMD_MEMLIMIT=y
 CONFIG_CMD_MEMSCAN=y
-# CONFIG_CMD_MEMTEST is not set
+CONFIG_CMD_MEMTEST=y
 
 #
 # Generic Drivers
@@ -117,7 +117,7 @@
 CONFIG_ENV=y
 CONFIG_ENV_LINK=y
 CONFIG_ENV_MUTABLE=y
-CONFIG_ENV_REGION="nor:0xee0000+15k"
+CONFIG_ENV_REGION="nor:0xfbc000+15k"
 # CONFIG_ENV_SAVEATONCE is not set
 CONFIG_ENV_CHECK_LEN=1024
 # CONFIG_VARIATIONS is not set
@@ -143,7 +143,7 @@
 #    Regions
 #
 CONFIG_ENV_REGION_KERNEL="fis://kernel"
-CONFIG_ENV_REGION_RAMDISK="fis://filesystem"
+CONFIG_ENV_REGION_RAMDISK="fis://ramdisk"
 
 #
 #    Overrides
Index: apex-1.5.13/src/mach-ixp42x/debian-nas100d-armeb_config
===================================================================
--- apex-1.5.13.orig/src/mach-ixp42x/debian-nas100d-armeb_config	2007-12-17 22:09:37.000000000 +1030
+++ apex-1.5.13/src/mach-ixp42x/debian-nas100d-armeb_config	2007-12-17 22:11:42.000000000 +1030
@@ -17,7 +17,7 @@
 #
 # General Setup
 #
-CONFIG_TARGET_DESCRIPTION="NAS100D"
+CONFIG_TARGET_DESCRIPTION="Debian NAS100D/BE"
 CONFIG_CROSS_COMPILE=""
 CONFIG_AEABI=y
 CONFIG_CC_OPTIMIZE_FOR_SIZE=y
@@ -92,7 +92,7 @@
 CONFIG_CMD_XRECEIVE=y
 CONFIG_CMD_MEMLIMIT=y
 CONFIG_CMD_MEMSCAN=y
-# CONFIG_CMD_MEMTEST is not set
+CONFIG_CMD_MEMTEST=y
 
 #
 # Generic Drivers
@@ -117,7 +117,7 @@
 CONFIG_ENV=y
 CONFIG_ENV_LINK=y
 CONFIG_ENV_MUTABLE=y
-CONFIG_ENV_REGION="nor:0xee0000+15k"
+CONFIG_ENV_REGION="nor:0xfbc000+15k"
 # CONFIG_ENV_SAVEATONCE is not set
 CONFIG_ENV_CHECK_LEN=1024
 # CONFIG_VARIATIONS is not set
@@ -134,16 +134,16 @@
 CONFIG_AUTOBOOT=y
 CONFIG_AUTOBOOT_DELAY=10
 CONFIG_ENV_STARTUP_KERNEL_COPY=y
-CONFIG_ENV_REGION_KERNEL_SWAP=y
+# CONFIG_ENV_REGION_KERNEL_SWAP is not set
 CONFIG_ENV_STARTUP_RAMDISK_COPY=y
-CONFIG_ENV_REGION_RAMDISK_SWAP=y
+# CONFIG_ENV_REGION_RAMDISK_SWAP is not set
 # CONFIG_ENV_STARTUP_PREFIX_P is not set
 
 #
 #    Regions
 #
 CONFIG_ENV_REGION_KERNEL="fis://kernel"
-CONFIG_ENV_REGION_RAMDISK="fis://filesystem"
+CONFIG_ENV_REGION_RAMDISK="fis://ramdisk"
 
 #
 #    Overrides
Index: apex-1.5.13/src/mach-ixp42x/slugos-nas100d-arm_config
===================================================================
--- apex-1.5.13.orig/src/mach-ixp42x/slugos-nas100d-arm_config	2007-12-17 22:09:37.000000000 +1030
+++ apex-1.5.13/src/mach-ixp42x/slugos-nas100d-arm_config	2007-12-17 22:11:55.000000000 +1030
@@ -17,7 +17,7 @@
 #
 # General Setup
 #
-CONFIG_TARGET_DESCRIPTION="SlugOS NAS100D"
+CONFIG_TARGET_DESCRIPTION="SlugOS NAS100D/LE"
 CONFIG_CROSS_COMPILE=""
 CONFIG_AEABI=y
 CONFIG_CC_OPTIMIZE_FOR_SIZE=y
@@ -90,7 +90,7 @@
 CONFIG_CMD_XRECEIVE=y
 CONFIG_CMD_MEMLIMIT=y
 CONFIG_CMD_MEMSCAN=y
-# CONFIG_CMD_MEMTEST is not set
+CONFIG_CMD_MEMTEST=y
 
 #
 # Generic Drivers
@@ -115,7 +115,7 @@
 CONFIG_ENV=y
 CONFIG_ENV_LINK=y
 CONFIG_ENV_MUTABLE=y
-CONFIG_ENV_REGION="nor:0xee0000+15k"
+CONFIG_ENV_REGION="nor:0xfbc000+15k"
 # CONFIG_ENV_SAVEATONCE is not set
 CONFIG_ENV_CHECK_LEN=1024
 # CONFIG_VARIATIONS is not set
Index: apex-1.5.13/src/mach-ixp42x/slugos-nas100d-armeb_config
===================================================================
--- apex-1.5.13.orig/src/mach-ixp42x/slugos-nas100d-armeb_config	2007-12-17 22:09:37.000000000 +1030
+++ apex-1.5.13/src/mach-ixp42x/slugos-nas100d-armeb_config	2007-12-17 22:12:08.000000000 +1030
@@ -17,7 +17,7 @@
 #
 # General Setup
 #
-CONFIG_TARGET_DESCRIPTION="SlugOS NAS100D (bigendian)"
+CONFIG_TARGET_DESCRIPTION="SlugOS NAS100D/BE"
 CONFIG_CROSS_COMPILE=""
 CONFIG_AEABI=y
 CONFIG_CC_OPTIMIZE_FOR_SIZE=y
@@ -90,7 +90,7 @@
 CONFIG_CMD_XRECEIVE=y
 CONFIG_CMD_MEMLIMIT=y
 CONFIG_CMD_MEMSCAN=y
-# CONFIG_CMD_MEMTEST is not set
+CONFIG_CMD_MEMTEST=y
 
 #
 # Generic Drivers
@@ -115,7 +115,7 @@
 CONFIG_ENV=y
 CONFIG_ENV_LINK=y
 CONFIG_ENV_MUTABLE=y
-CONFIG_ENV_REGION="nor:0xee0000+15k"
+CONFIG_ENV_REGION="nor:0xfbc000+15k"
 # CONFIG_ENV_SAVEATONCE is not set
 CONFIG_ENV_CHECK_LEN=1024
 # CONFIG_VARIATIONS is not set
