From: Rod Whitby <rod@whitby.id.au>
Subject: Debian support for the IOMega NAS100D

Config files for running Debian on the NAS100D.

Signed-off-by: Rod Whitby <rod@whitby.id.au>
--

Index: apex/src/mach-ixp42x/debian-nas100d-arm_config
===================================================================
--- apex.orig/src/mach-ixp42x/debian-nas100d-arm_config	2008-05-28 10:52:50.000000000 +0930
+++ apex/src/mach-ixp42x/debian-nas100d-arm_config	2008-05-28 10:52:56.000000000 +0930
@@ -17,7 +17,7 @@
 #
 # General Setup
 #
-CONFIG_TARGET_DESCRIPTION="NAS100D"
+CONFIG_TARGET_DESCRIPTION="Debian NAS100D/LE"
 CONFIG_CROSS_COMPILE=""
 CONFIG_AEABI=y
 CONFIG_CC_OPTIMIZE_FOR_SIZE=y
@@ -92,7 +92,7 @@
 CONFIG_CMD_XRECEIVE=y
 CONFIG_CMD_MEMLIMIT=y
 CONFIG_CMD_MEMSCAN=y
-# CONFIG_CMD_MEMTEST is not set
+CONFIG_CMD_MEMTEST=y
 
 #
 # Generic Drivers
@@ -117,7 +117,7 @@
 CONFIG_ENV=y
 CONFIG_ENV_LINK=y
 CONFIG_ENV_MUTABLE=y
-CONFIG_ENV_REGION="nor:0xee0000+15k"
+CONFIG_ENV_REGION="nor:0xfbc000+15k"
 # CONFIG_ENV_SAVEATONCE is not set
 CONFIG_ENV_CHECK_LEN=1024
 # CONFIG_VARIATIONS is not set
@@ -143,7 +143,7 @@
 #    Regions
 #
 CONFIG_ENV_REGION_KERNEL="fis://kernel"
-CONFIG_ENV_REGION_RAMDISK="fis://filesystem"
+CONFIG_ENV_REGION_RAMDISK="fis://ramdisk"
 
 #
 #    Overrides
Index: apex/src/mach-ixp42x/debian-nas100d-armeb_config
===================================================================
--- apex.orig/src/mach-ixp42x/debian-nas100d-armeb_config	2008-05-28 10:52:50.000000000 +0930
+++ apex/src/mach-ixp42x/debian-nas100d-armeb_config	2008-05-28 10:52:56.000000000 +0930
@@ -17,7 +17,7 @@
 #
 # General Setup
 #
-CONFIG_TARGET_DESCRIPTION="NAS100D"
+CONFIG_TARGET_DESCRIPTION="Debian NAS100D/BE"
 CONFIG_CROSS_COMPILE=""
 CONFIG_AEABI=y
 CONFIG_CC_OPTIMIZE_FOR_SIZE=y
@@ -92,7 +92,7 @@
 CONFIG_CMD_XRECEIVE=y
 CONFIG_CMD_MEMLIMIT=y
 CONFIG_CMD_MEMSCAN=y
-# CONFIG_CMD_MEMTEST is not set
+CONFIG_CMD_MEMTEST=y
 
 #
 # Generic Drivers
@@ -117,7 +117,7 @@
 CONFIG_ENV=y
 CONFIG_ENV_LINK=y
 CONFIG_ENV_MUTABLE=y
-CONFIG_ENV_REGION="nor:0xee0000+15k"
+CONFIG_ENV_REGION="nor:0xfbc000+15k"
 # CONFIG_ENV_SAVEATONCE is not set
 CONFIG_ENV_CHECK_LEN=1024
 # CONFIG_VARIATIONS is not set
@@ -134,16 +134,16 @@
 CONFIG_AUTOBOOT=y
 CONFIG_AUTOBOOT_DELAY=10
 CONFIG_ENV_STARTUP_KERNEL_COPY=y
-CONFIG_ENV_REGION_KERNEL_SWAP=y
+# CONFIG_ENV_REGION_KERNEL_SWAP is not set
 CONFIG_ENV_STARTUP_RAMDISK_COPY=y
-CONFIG_ENV_REGION_RAMDISK_SWAP=y
+# CONFIG_ENV_REGION_RAMDISK_SWAP is not set
 # CONFIG_ENV_STARTUP_PREFIX_P is not set
 
 #
 #    Regions
 #
 CONFIG_ENV_REGION_KERNEL="fis://kernel"
-CONFIG_ENV_REGION_RAMDISK="fis://filesystem"
+CONFIG_ENV_REGION_RAMDISK="fis://ramdisk"
 
 #
 #    Overrides
Index: apex/src/mach-ixp42x/slugos-nas100d-arm_config
===================================================================
--- apex.orig/src/mach-ixp42x/slugos-nas100d-arm_config	2008-05-28 10:52:50.000000000 +0930
+++ apex/src/mach-ixp42x/slugos-nas100d-arm_config	2008-05-28 10:52:56.000000000 +0930
@@ -17,7 +17,7 @@
 #
 # General Setup
 #
-CONFIG_TARGET_DESCRIPTION="SlugOS NAS100D"
+CONFIG_TARGET_DESCRIPTION="SlugOS NAS100D/LE"
 CONFIG_CROSS_COMPILE=""
 CONFIG_AEABI=y
 CONFIG_CC_OPTIMIZE_FOR_SIZE=y
@@ -90,7 +90,7 @@
 CONFIG_CMD_XRECEIVE=y
 CONFIG_CMD_MEMLIMIT=y
 CONFIG_CMD_MEMSCAN=y
-# CONFIG_CMD_MEMTEST is not set
+CONFIG_CMD_MEMTEST=y
 
 #
 # Generic Drivers
@@ -115,7 +115,7 @@
 CONFIG_ENV=y
 CONFIG_ENV_LINK=y
 CONFIG_ENV_MUTABLE=y
-CONFIG_ENV_REGION="nor:0xee0000+15k"
+CONFIG_ENV_REGION="nor:0xfbc000+15k"
 # CONFIG_ENV_SAVEATONCE is not set
 CONFIG_ENV_CHECK_LEN=1024
 # CONFIG_VARIATIONS is not set
Index: apex/src/mach-ixp42x/slugos-nas100d-armeb_config
===================================================================
--- apex.orig/src/mach-ixp42x/slugos-nas100d-armeb_config	2008-05-28 10:52:50.000000000 +0930
+++ apex/src/mach-ixp42x/slugos-nas100d-armeb_config	2008-05-28 10:52:56.000000000 +0930
@@ -17,7 +17,7 @@
 #
 # General Setup
 #
-CONFIG_TARGET_DESCRIPTION="SlugOS NAS100D (bigendian)"
+CONFIG_TARGET_DESCRIPTION="SlugOS NAS100D/BE"
 CONFIG_CROSS_COMPILE=""
 CONFIG_AEABI=y
 CONFIG_CC_OPTIMIZE_FOR_SIZE=y
@@ -90,7 +90,7 @@
 CONFIG_CMD_XRECEIVE=y
 CONFIG_CMD_MEMLIMIT=y
 CONFIG_CMD_MEMSCAN=y
-# CONFIG_CMD_MEMTEST is not set
+CONFIG_CMD_MEMTEST=y
 
 #
 # Generic Drivers
@@ -115,7 +115,7 @@
 CONFIG_ENV=y
 CONFIG_ENV_LINK=y
 CONFIG_ENV_MUTABLE=y
-CONFIG_ENV_REGION="nor:0xee0000+15k"
+CONFIG_ENV_REGION="nor:0xfbc000+15k"
 # CONFIG_ENV_SAVEATONCE is not set
 CONFIG_ENV_CHECK_LEN=1024
 # CONFIG_VARIATIONS is not set
