From: Rod Whitby <rod@whitby.id.au>
Subject: Correct SlugOS nslu2 apex environment locations

Move the apex environment into the apex partition.

Signed-off-by: Rod Whitby <rod@whitby.id.au>
--

Index: apex/src/mach-ixp42x/slugos-nslu2-16mb-arm_config
===================================================================
--- apex.orig/src/mach-ixp42x/slugos-nslu2-16mb-arm_config	2008-05-28 10:52:58.000000000 +0930
+++ apex/src/mach-ixp42x/slugos-nslu2-16mb-arm_config	2008-05-28 10:53:04.000000000 +0930
@@ -117,7 +117,7 @@
 CONFIG_ENV=y
 CONFIG_ENV_LINK=y
 CONFIG_ENV_MUTABLE=y
-CONFIG_ENV_REGION="nor:0x7c000+16k"
+CONFIG_ENV_REGION="nor:0x7c000+15k"
 # CONFIG_ENV_SAVEATONCE is not set
 CONFIG_ENV_CHECK_LEN=1024
 # CONFIG_VARIATIONS is not set
Index: apex/src/mach-ixp42x/slugos-nslu2-16mb-armeb_config
===================================================================
--- apex.orig/src/mach-ixp42x/slugos-nslu2-16mb-armeb_config	2008-05-28 10:52:50.000000000 +0930
+++ apex/src/mach-ixp42x/slugos-nslu2-16mb-armeb_config	2008-05-28 10:53:04.000000000 +0930
@@ -117,7 +117,7 @@
 CONFIG_ENV=y
 CONFIG_ENV_LINK=y
 CONFIG_ENV_MUTABLE=y
-CONFIG_ENV_REGION="nor:0x7c000+16k"
+CONFIG_ENV_REGION="nor:0x7c000+15k"
 # CONFIG_ENV_SAVEATONCE is not set
 CONFIG_ENV_CHECK_LEN=1024
 # CONFIG_VARIATIONS is not set
Index: apex/src/mach-ixp42x/slugos-nslu2-arm_config
===================================================================
--- apex.orig/src/mach-ixp42x/slugos-nslu2-arm_config	2008-05-28 10:53:00.000000000 +0930
+++ apex/src/mach-ixp42x/slugos-nslu2-arm_config	2008-05-28 10:53:04.000000000 +0930
@@ -117,7 +117,7 @@
 CONFIG_ENV=y
 CONFIG_ENV_LINK=y
 CONFIG_ENV_MUTABLE=y
-CONFIG_ENV_REGION="nor:0x7f8000+16k"
+CONFIG_ENV_REGION="nor:0x7c000+15k"
 # CONFIG_ENV_SAVEATONCE is not set
 CONFIG_ENV_CHECK_LEN=1024
 # CONFIG_VARIATIONS is not set
Index: apex/src/mach-ixp42x/slugos-nslu2-armeb_config
===================================================================
--- apex.orig/src/mach-ixp42x/slugos-nslu2-armeb_config	2008-05-28 10:53:00.000000000 +0930
+++ apex/src/mach-ixp42x/slugos-nslu2-armeb_config	2008-05-28 10:53:04.000000000 +0930
@@ -117,7 +117,7 @@
 CONFIG_ENV=y
 CONFIG_ENV_LINK=y
 CONFIG_ENV_MUTABLE=y
-CONFIG_ENV_REGION="nor:0x7f8000+16k"
+CONFIG_ENV_REGION="nor:0x7c000+15k"
 # CONFIG_ENV_SAVEATONCE is not set
 CONFIG_ENV_CHECK_LEN=1024
 # CONFIG_VARIATIONS is not set
