Index: linux-2.6.20.1.orig/arch/ppc/platforms/sandpoint.h
===================================================================
--- linux-2.6.20.1.orig.orig/arch/ppc/platforms/sandpoint.h	2007-02-23 17:20:30.000000000 -0500
+++ linux-2.6.20.1.orig/arch/ppc/platforms/sandpoint.h	2007-02-23 17:26:13.000000000 -0500
@@ -49,29 +49,45 @@
 /*
  * Serial defines.
  */
-#define SANDPOINT_SERIAL_0		0xfe0003f8
-#define SANDPOINT_SERIAL_1		0xfe0002f8
+#define SANDPOINT_SERIAL_0		0xfdf04500
+#define SANDPOINT_SERIAL_1		0xfdf04600
 
-#define RS_TABLE_SIZE  2
+#define RS_TABLE_SIZE  1
 
 /* Rate for the 1.8432 Mhz clock for the onboard serial chip */
-#define BASE_BAUD			( 1843200 / 16 )
-#define UART_CLK			1843200
+#ifdef CONFIG_QNAP_UART_CLK_100
+#define BASE_BAUD			( 100000000 / 16 )
+#define UART_CLK			100000000
+#else
+#define BASE_BAUD			( 133000000 / 16 )
+#define UART_CLK			133000000
+#endif
 
 #ifdef CONFIG_SERIAL_DETECT_IRQ
-#define STD_COM_FLAGS (ASYNC_BOOT_AUTOCONF|ASYNC_AUTO_IRQ)
+#define STD_COM_FLAGS (ASYNC_BOOT_AUTOCONF|ASYNC_AUTO_IRQ|ASYNC_SKIP_TEST)
 #else
-#define STD_COM_FLAGS (ASYNC_BOOT_AUTOCONF)
+#define STD_COM_FLAGS (ASYNC_BOOT_AUTOCONF|ASYNC_SKIP_TEST)
 #endif
 
+#define SANDPOINT_SERIAL_0_INT 25 // 4 + EPIC_IRQ_BASE + NUM_8259_INTERRUPTS
+#define SANDPOINT_SERIAL_1_INT 24 // 3 + EPIC_IRQ_BASE + NUM_8259_INTERRUPTS
+#if 0
+
 #define STD_SERIAL_PORT_DFNS \
-        { 0, BASE_BAUD, SANDPOINT_SERIAL_0, 4, STD_COM_FLAGS, /* ttyS0 */ \
+        { 0, BASE_BAUD, SANDPOINT_SERIAL_0, SANDPOINT_SERIAL_0_INT, STD_COM_FLAGS, /* ttyS0 */ \
 		iomem_base: (u8 *)SANDPOINT_SERIAL_0,			  \
 		io_type: SERIAL_IO_MEM },				  \
-        { 0, BASE_BAUD, SANDPOINT_SERIAL_1, 3, STD_COM_FLAGS, /* ttyS1 */ \
+        { 0, BASE_BAUD, SANDPOINT_SERIAL_1, SANDPOINT_SERIAL_1_INT, STD_COM_FLAGS, /* ttyS1 */ \
 		iomem_base: (u8 *)SANDPOINT_SERIAL_1,			  \
 		io_type: SERIAL_IO_MEM },
 
+#else
+#define STD_SERIAL_PORT_DFNS \
+        { 0, BASE_BAUD, SANDPOINT_SERIAL_0, SANDPOINT_SERIAL_0_INT, STD_COM_FLAGS, /* ttyS0 */ \
+                iomem_base: (u8 *)SANDPOINT_SERIAL_0,                     \
+                io_type: SERIAL_IO_MEM },
+#endif
+
 #define SERIAL_PORT_DFNS \
         STD_SERIAL_PORT_DFNS
 
Index: linux-2.6.20.1.orig/arch/ppc/platforms/qnap_pic.c
===================================================================
--- linux-2.6.20.1.orig.orig/arch/ppc/platforms/qnap_pic.c	2007-02-23 17:30:16.000000000 -0500
+++ linux-2.6.20.1.orig/arch/ppc/platforms/qnap_pic.c	2007-02-23 17:30:20.000000000 -0500
@@ -1,4 +1,3 @@
-#include <linux/config.h>
 #include <linux/module.h>
 #include <linux/poll.h>
 #include <linux/miscdevice.h>
Index: linux-2.6.20.1.orig/arch/ppc/syslib/Makefile
===================================================================
--- linux-2.6.20.1.orig.orig/arch/ppc/syslib/Makefile	2007-02-23 17:37:37.000000000 -0500
+++ linux-2.6.20.1.orig/arch/ppc/syslib/Makefile	2007-02-23 17:37:47.000000000 -0500
@@ -66,7 +66,7 @@
 obj-$(CONFIG_HARRIER)		+= harrier.o
 obj-$(CONFIG_PRPMC800)		+= open_pic.o pci_auto.o
 obj-$(CONFIG_RADSTONE_PPC7D)	+= pci_auto.o
-obj-$(CONFIG_SANDPOINT)		+= pci_auto.o
+obj-$(CONFIG_SANDPOINT)		+= pci_auto.o todc_time.o
 obj-$(CONFIG_SBC82xx)		+= todc_time.o
 obj-$(CONFIG_SPRUCE)		+= cpc700_pic.o pci_auto.o \
 				   todc_time.o
