Index: linux-2.6.17/drivers/scsi/sd.c
===================================================================
--- linux-2.6.17.orig/drivers/scsi/sd.c	2006-09-05 00:52:10.000000000 +0200
+++ linux-2.6.17/drivers/scsi/sd.c	2006-09-05 01:16:03.000000000 +0200
@@ -376,6 +376,9 @@
 			    "count=%d\n", disk->disk_name,
 			 (unsigned long long)block, this_count));
 
+	/* Update idle-since time */
+	sdp->idle = jiffies;
+
 	if (!sdp || !scsi_device_online(sdp) ||
  	    block + rq->nr_sectors > get_capacity(disk)) {
 		SCSI_LOG_HLQUEUE(2, printk("Finishing %ld sectors\n", 
@@ -705,6 +708,8 @@
 		case SCSI_IOCTL_GET_IDLUN:
 		case SCSI_IOCTL_GET_BUS_NUMBER:
 			return scsi_ioctl(sdp, cmd, p);
+		case SCSI_IOCTL_IDLE:
+			return (jiffies - sdp->idle) / HZ + 1;
 		default:
 			error = scsi_cmd_ioctl(filp, disk, cmd, p);
 			if (error != -ENOTTY)
Index: linux-2.6.17/include/scsi/scsi_device.h
===================================================================
--- linux-2.6.17.orig/include/scsi/scsi_device.h	2006-09-05 00:33:21.000000000 +0200
+++ linux-2.6.17/include/scsi/scsi_device.h	2006-09-05 01:16:03.000000000 +0200
@@ -134,6 +134,7 @@
 	atomic_t ioerr_cnt;
 
 	int timeout;
+	unsigned long		idle;	/* scsi idle time in jiffers */
 
 	struct device		sdev_gendev;
 	struct class_device	sdev_classdev;
Index: linux-2.6.17/include/scsi/scsi_ioctl.h
===================================================================
--- linux-2.6.17.orig/include/scsi/scsi_ioctl.h	2006-09-05 00:33:21.000000000 +0200
+++ linux-2.6.17/include/scsi/scsi_ioctl.h	2006-09-05 01:16:03.000000000 +0200
@@ -7,6 +7,7 @@
 #define SCSI_IOCTL_SYNC 4			/* Request synchronous parameters */
 #define SCSI_IOCTL_START_UNIT 5
 #define SCSI_IOCTL_STOP_UNIT 6
+#define SCSI_IOCTL_IDLE 4746			/* get idle time in jiffies */
 /* The door lock/unlock constants are compatible with Sun constants for
    the cdrom */
 #define SCSI_IOCTL_DOORLOCK 0x5380		/* lock the eject mechanism */
