Index: linux-2.6.24-arm/arch/arm/mach-ixp4xx/fsg-setup.c
===================================================================
--- linux-2.6.24-arm.orig/arch/arm/mach-ixp4xx/fsg-setup.c	2008-01-29 23:06:24.000000000 +1030
+++ linux-2.6.24-arm/arch/arm/mach-ixp4xx/fsg-setup.c	2008-01-29 23:04:32.000000000 +1030
@@ -15,6 +15,7 @@
 #include <linux/kernel.h>
 #include <linux/serial.h>
 #include <linux/serial_8250.h>
+#include <linux/leds.h>
 #include <linux/reboot.h>
 #include <linux/i2c-gpio.h>
 
@@ -96,6 +97,11 @@
 	.resource		= fsg_uart_resources,
 };
 
+static struct platform_device fsg_leds = {
+	.name		= "fsg-led",
+	.id		= -1,
+};
+
 /* Built-in 10/100 Ethernet MAC interfaces */
 static struct eth_plat_info fsg_plat_eth[] = {
 	{
@@ -124,6 +130,7 @@
 static struct platform_device *fsg_devices[] __initdata = {
 	&fsg_i2c_gpio,
 	&fsg_flash,
+	&fsg_leds,
 	&fsg_eth[0],
 	&fsg_eth[1],
 };
