Index: linux-2.6.21-rc4-arm/arch/arm/mach-ixp4xx/dsmg600-setup.c
===================================================================
--- linux-2.6.21-rc4-arm.orig/arch/arm/mach-ixp4xx/dsmg600-setup.c	2007-03-21 02:37:02.000000000 -0800
+++ linux-2.6.21-rc4-arm/arch/arm/mach-ixp4xx/dsmg600-setup.c	2007-03-21 02:37:34.000000000 -0800
@@ -15,6 +15,7 @@
 #include <linux/serial.h>
 #include <linux/serial_8250.h>
 
+#include <asm/setup.h>
 #include <asm/mach-types.h>
 #include <asm/mach/arch.h>
 #include <asm/mach/flash.h>
@@ -156,9 +157,35 @@
 #endif
 }
 
+static char dsmg600_rtc_probe[] __initdata = "rtc-pcf8563.probe=0,0x51 ";
+
 static void __init dsmg600_fixup(struct machine_desc *desc,
                 struct tag *tags, char **cmdline, struct meminfo *mi)
 {
+        struct tag *t = tags;
+        char *p = *cmdline;
+        
+        /* Find the end of the tags table, taking note of any cmdline tag. */
+        for (; t->hdr.size; t = tag_next(t)) {
+                if (t->hdr.tag == ATAG_CMDLINE) {
+                        p = t->u.cmdline.cmdline;
+                }
+        }       
+
+        /* Overwrite the end of the table with a new cmdline tag. */
+        t->hdr.tag = ATAG_CMDLINE;
+        t->hdr.size =
+                (sizeof (struct tag_header) +
+                 strlen(dsmg600_rtc_probe) + strlen(p) + 1 + 4) >> 2;
+        strlcpy(t->u.cmdline.cmdline, dsmg600_rtc_probe, COMMAND_LINE_SIZE);
+        strlcpy(t->u.cmdline.cmdline + strlen(dsmg600_rtc_probe), p,
+                COMMAND_LINE_SIZE - strlen(dsmg600_rtc_probe));
+                
+        /* Terminate the table. */
+        t = tag_next(t);
+        t->hdr.tag = ATAG_NONE;                        
+        t->hdr.size = 0;
+
        /* The xtal on this machine is non-standard. */
         ixp4xx_timer_freq = DSMG600_FREQ;
 }
