Index: linux-2.6.23.12-armeb/arch/arm/mach-ixp4xx/dsmg600-setup.c
===================================================================
--- linux-2.6.23.12-armeb.orig/arch/arm/mach-ixp4xx/dsmg600-setup.c	2008-01-12 11:58:32.000000000 +1030
+++ linux-2.6.23.12-armeb/arch/arm/mach-ixp4xx/dsmg600-setup.c	2008-01-12 12:00:37.000000000 +1030
@@ -14,6 +14,7 @@
 #include <linux/kernel.h>
 #include <linux/serial.h>
 #include <linux/serial_8250.h>
+#include <linux/i2c.h>
 #include <linux/i2c-gpio.h>
 
 #include <asm/mach-types.h>
@@ -51,6 +52,12 @@
 	},
 };
 
+static struct i2c_board_info __initdata dsmg600_i2c_board_info [] = {
+	{
+		I2C_BOARD_INFO("rtc-pcf8563", 0x51),
+	},
+};
+
 #ifdef CONFIG_LEDS_CLASS
 static struct resource dsmg600_led_resources[] = {
 	{
@@ -166,6 +173,9 @@
 
 	platform_add_devices(dsmg600_devices, ARRAY_SIZE(dsmg600_devices));
 
+	i2c_register_board_info(0, dsmg600_i2c_board_info,
+				ARRAY_SIZE(dsmg600_i2c_board_info));
+
 #ifdef CONFIG_LEDS_CLASS
         /* We don't care whether or not this works. */
         (void)platform_device_register(&dsmg600_leds);
