diff --git a/drivers/net/ixp4xx/ucode_dl.c b/drivers/net/ixp4xx/ucode_dl.c
index 3807320..b112236 100644
--- a/drivers/net/ixp4xx/ucode_dl.c
+++ b/drivers/net/ixp4xx/ucode_dl.c
@@ -197,11 +197,10 @@ static ssize_t ucode_write(struct file *
 		char *data;
 		struct dl_image *image;
 	} u;
+	const char __user *cbuf = buf;
 
 	u.data = file->private_data;
 
-	const char __user *cbuf = buf;
-
 	while (count) {
 		int len;
 		if (*ppos < sizeof(struct dl_image)) {
@@ -396,7 +395,7 @@ static int npe_remove(struct platform_de
 }
 
 static struct platform_driver ixp4xx_npe_driver = {
-	.driver {
+	.driver = {
 		.name	= "ixp4xx_npe",
 		.owner	= THIS_MODULE,
 	},
