--- linux-2.6.16/arch/arm/common/dmabounce.c~	2006-03-19 23:53:29.000000000 -0600
+++ linux-2.6.16/arch/arm/common/dmabounce.c	2006-06-12 11:59:39.000000000 -0500
@@ -686,6 +686,7 @@
 EXPORT_SYMBOL(dma_map_sg);
 EXPORT_SYMBOL(dma_unmap_sg);
 EXPORT_SYMBOL(dma_sync_single);
+EXPORT_SYMBOL(dma_sync_single_for_device);
 EXPORT_SYMBOL(dma_sync_sg);
 EXPORT_SYMBOL(dmabounce_register_dev);
 EXPORT_SYMBOL(dmabounce_unregister_dev);
