Index: linux-2.6.18-rc6/arch/arm/mach-ixp4xx/nslu2-pci.c
===================================================================
--- linux-2.6.18-rc6.orig/arch/arm/mach-ixp4xx/nslu2-pci.c	2006-09-11 21:42:24.000000000 +0200
+++ linux-2.6.18-rc6/arch/arm/mach-ixp4xx/nslu2-pci.c	2006-09-11 21:44:50.000000000 +0200
@@ -17,6 +17,7 @@
 
 #include <linux/pci.h>
 #include <linux/init.h>
+#include <linux/irq.h>
 
 #include <asm/mach/pci.h>
 #include <asm/mach-types.h>
Index: linux-2.6.18-rc6/arch/arm/mach-ixp4xx/nslu2-power.c
===================================================================
--- linux-2.6.18-rc6.orig/arch/arm/mach-ixp4xx/nslu2-power.c	2006-09-11 21:42:24.000000000 +0200
+++ linux-2.6.18-rc6/arch/arm/mach-ixp4xx/nslu2-power.c	2006-09-11 21:45:24.000000000 +0200
@@ -21,6 +21,7 @@
 #include <linux/reboot.h>
 #include <linux/interrupt.h>
 #include <linux/reboot.h>
+#include <linux/irq.h>
 
 #include <asm/mach-types.h>
 
