Index: linux-2.6.19/drivers/net/ixp4xx/mac_driver.c
===================================================================
--- linux-2.6.19.orig/drivers/net/ixp4xx/mac_driver.c	2007-01-02 18:53:24.420052213 +0000
+++ linux-2.6.19/drivers/net/ixp4xx/mac_driver.c	2007-01-02 18:57:59.973273214 +0000
@@ -434,9 +434,14 @@
 	 * following commands:
 	 * "ip link set address 02:03:04:04:04:01 dev eth0"
 	 * "ifconfig eth0 hw ether 02:03:04:04:04:07"
-	 */
-	random_ether_addr(dev->dev_addr);
-	dev->dev_addr[5] = plat->phy_id;
+        */
+
+        if (is_zero_ether_addr(plat->hwaddr)) {
+                random_ether_addr(dev->dev_addr);
+                dev->dev_addr[5] = plat->phy_id;
+        }
+        else
+                memcpy(dev->dev_addr, plat->hwaddr, 6);
 
 	printk(KERN_INFO IXMAC_NAME " driver " IXMAC_VERSION
 			": %s on %s with PHY[%d] initialized\n",
Index: linux-2.6.19/include/asm-arm/arch-ixp4xx/platform.h
===================================================================
--- linux-2.6.19.orig/include/asm-arm/arch-ixp4xx/platform.h	2007-01-02 18:53:24.460054713 +0000
+++ linux-2.6.19/include/asm-arm/arch-ixp4xx/platform.h	2007-01-02 18:56:43.412488467 +0000
@@ -100,6 +100,7 @@
 	int phy_id;	/* ID of the connected PHY (PCB/platform dependent) */
 	int rxq_id;	/* Queue ID of the RX-free q*/
 	int txq_id;	/* Where to push the outgoing packets */
+        unsigned char hwaddr[6]; /* Desired hardware address */
 };
 
 
