--- linux-2.6.20/arch/arm/mach-ixp4xx/dsmg600-setup.c.orig	2007-03-05 22:42:00.000000000 -0600
+++ linux-2.6.20/arch/arm/mach-ixp4xx/dsmg600-setup.c	2007-03-05 22:39:58.000000000 -0600
@@ -132,7 +132,6 @@
 
 static void __init dsmg600_init(void)
 {
-        ixp4xx_timer_freq = DSMG600_FREQ;
 	ixp4xx_sys_init();
 
 	/* Make sure that GPIO14 and GPIO15 are not used as clocks */
@@ -186,6 +185,9 @@
 	t = tag_next(t);
 	t->hdr.tag = ATAG_NONE;
 	t->hdr.size = 0;
+
+	/* The xtal on this machine is non-standard. */
+        ixp4xx_timer_freq = DSMG600_FREQ;
 }
 
 MACHINE_START(DSMG600, "D-Link DSM-G600 RevA")
