Index: linux-2.6.22-rc1-arm/arch/arm/Kconfig
===================================================================
--- linux-2.6.22-rc1-arm.orig/arch/arm/Kconfig	2007-05-12 23:31:31.000000000 -0700
+++ linux-2.6.22-rc1-arm/arch/arm/Kconfig	2007-05-12 23:37:25.000000000 -0700
@@ -287,6 +287,7 @@
 config ARCH_IXP4XX
 	bool "IXP4xx-based"
 	depends on MMU
+	select GENERIC_GPIO
 	select GENERIC_TIME
 	select GENERIC_CLOCKEVENTS
 	help
