Index: linux-2.6.16/arch/arm/mm/copypage-xscale.c
===================================================================
--- linux-2.6.16.orig/arch/arm/mm/copypage-xscale.c	2006-06-07 23:24:52.098025512 +0930
+++ linux-2.6.16/arch/arm/mm/copypage-xscale.c	2006-06-07 23:25:30.853133840 +0930
@@ -15,6 +15,7 @@
  */
 #include <linux/init.h>
 #include <linux/mm.h>
+#include <linux/module.h>
 
 #include <asm/page.h>
 #include <asm/pgtable.h>
@@ -129,3 +130,6 @@
 	.cpu_clear_user_page	= xscale_mc_clear_user_page, 
 	.cpu_copy_user_page	= xscale_mc_copy_user_page,
 };
+
+EXPORT_SYMBOL(xscale_mc_clear_user_page);
+EXPORT_SYMBOL(xscale_mc_copy_user_page);
