---
 Makefile |    3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

Index: arm-kernel-shim-1.2/Makefile
===================================================================
--- arm-kernel-shim-1.2.orig/Makefile
+++ arm-kernel-shim-1.2/Makefile
@@ -9,7 +9,8 @@
 OBJCOPY=$(CROSS_COMPILE)objcopy
 
 CFLAGS:=-g -Os -nostartfiles -nostdlib
-LDFLAGS:= -g -X -Map shim.map -Ttext 0
+# LDFLAGS:= -g -X -Map shim.map -Ttext 0
+LDFLAGS:= -g -X -Map shim.map -Ttext 0 --section-start .boot=1fb
 
 SRCS:= shim.lds main.c
 OBJS:=$(SRCS:.c=.o)
