Index: linux-2.6.16/arch/arm/mach-ixp4xx/common.c
===================================================================
--- linux-2.6.16.orig/arch/arm/mach-ixp4xx/common.c	2006-06-07 23:24:51.917053024 +0930
+++ linux-2.6.16/arch/arm/mach-ixp4xx/common.c	2006-06-07 23:25:31.358057080 +0930
@@ -91,7 +91,7 @@
 /*
  * IRQ -> GPIO mapping table
  */
-static int irq2gpio[32] = {
+static char irq2gpio[32] = {
 	-1, -1, -1, -1, -1, -1,  0,  1,
 	-1, -1, -1, -1, -1, -1, -1, -1,
 	-1, -1, -1,  2,  3,  4,  5,  6,
@@ -153,6 +153,9 @@
 	/* Set the new style */
 	*int_reg |= (int_style << (line * IXP4XX_GPIO_STYLE_SIZE));
 
+	/* Configure the line as an input */
+	gpio_line_config(line, IXP4XX_GPIO_IN);
+
 	return 0;
 }
 
