Index: linux-2.6.24-rc7-armeb/arch/arm/mach-ixp4xx/nas100d-setup.c
===================================================================
--- linux-2.6.24-rc7-armeb.orig/arch/arm/mach-ixp4xx/nas100d-setup.c	2008-01-14 15:49:05.000000000 +1030
+++ linux-2.6.24-rc7-armeb/arch/arm/mach-ixp4xx/nas100d-setup.c	2008-01-14 15:49:36.000000000 +1030
@@ -16,6 +16,7 @@
 #include <linux/serial.h>
 #include <linux/serial_8250.h>
 #include <linux/leds.h>
+#include <linux/i2c.h>
 #include <linux/i2c-gpio.h>
 
 #include <asm/mach-types.h>
@@ -39,6 +40,12 @@
 	.resource		= &nas100d_flash_resource,
 };
 
+static struct i2c_board_info __initdata nas100d_i2c_board_info [] = {
+	{
+		I2C_BOARD_INFO("rtc-pcf8563", 0x51),
+	},
+};
+
 #ifdef CONFIG_LEDS_IXP4XX
 static struct resource nas100d_led_resources[] = {
 	{
@@ -157,6 +164,9 @@
 
 	pm_power_off = nas100d_power_off;
 
+	i2c_register_board_info(0, nas100d_i2c_board_info,
+				ARRAY_SIZE(nas100d_i2c_board_info));
+
 	/*
 	 * This is only useful on a modified machine, but it is valuable
 	 * to have it first in order to see debug messages, and so that
