Index: linux-2.6.17-rc6/arch/arm/mach-ixp4xx/common.c
===================================================================
--- linux-2.6.17-rc6.orig/arch/arm/mach-ixp4xx/common.c	2006-06-07 23:50:22.415381912 +0930
+++ linux-2.6.17-rc6/arch/arm/mach-ixp4xx/common.c	2006-06-07 23:59:28.813316800 +0930
@@ -156,6 +156,9 @@
 	/* Configure the line as an input */
 	gpio_line_config(line, IXP4XX_GPIO_IN);
 
+	/* Configure the line as an input */
+	gpio_line_config(line, IXP4XX_GPIO_IN);
+
 	return 0;
 }
 
