diff -Naur linux-2.6.18-rc6.orig/drivers/net/wireless/zd1211rw/zd_usb.c linux-2.6.18-rc6/drivers/net/wireless/zd1211rw/zd_usb.c
--- linux-2.6.18-rc6.orig/drivers/net/wireless/zd1211rw/zd_usb.c	2006-09-12 20:41:03.000000000 +0100
+++ linux-2.6.18-rc6/drivers/net/wireless/zd1211rw/zd_usb.c	2006-09-12 20:42:36.000000000 +0100
@@ -527,7 +527,7 @@
 {
 	int i;
 	struct zd_mac *mac = zd_usb_to_mac(usb);
-	const struct rx_length_info *length_info;
+	struct rx_length_info *length_info;
 
 	if (length < sizeof(struct rx_length_info)) {
 		/* It's not a complete packet anyhow. */
