diff -Naur linux-2.6.18.orig/arch/arm/mach-ixp4xx/nas100d-setup.c linux-2.6.18/arch/arm/mach-ixp4xx/nas100d-setup.c
--- linux-2.6.18.orig/arch/arm/mach-ixp4xx/nas100d-setup.c	2006-09-28 14:59:01.790938000 +0000
+++ linux-2.6.18/arch/arm/mach-ixp4xx/nas100d-setup.c	2006-09-28 14:55:48.622865750 +0000
@@ -128,12 +128,35 @@
 	.resource		= nas100d_uart_resources,
 };
 
+static struct resource res_mac0 = {
+       .start          = IXP4XX_EthB_BASE_PHYS,
+       .end            = IXP4XX_EthB_BASE_PHYS + 0x1ff,
+       .flags          = IORESOURCE_MEM,
+};
+
+static struct mac_plat_info plat_mac0 = {
+       .npe_id         = 1,
+       .phy_id         = 0,
+       .eth_id         = 0,
+       .rxq_id         = 27,
+       .txq_id         = 24,
+};
+
+static struct platform_device mac0 = {
+       .name           = "ixp4xx_mac",
+       .id             = 0,
+       .dev.platform_data = &plat_mac0,
+       .num_resources  = 1,
+       .resource       = &res_mac0,
+};
+
 static struct platform_device *nas100d_devices[] __initdata = {
 	&nas100d_i2c_controller,
 	&nas100d_flash,
 #ifdef CONFIG_LEDS_IXP4XX
 	&nas100d_leds,
 #endif
+	&mac0
 };
 
 static void nas100d_power_off(void)
